{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Using predictors trained on DMSexp and EVmutation datasets\n",
    "\n",
    "This vignette shows how to prepare variants for applying our predictors trained on DMSexp and EVmutation datasets. \n",
    "\n",
    "For illustration we will use the BRCA1 DMS experimental dataset by [Findlay et al Nature 2018](https://www.nature.com/articles/s41586-018-0461-z). The code (python and R) below shows how to annotate the features required for the predictors, and applying the classifiers. We will have the actual experimental data, as well as predictions offered by [EVmutation](https://marks.hms.harvard.edu/evmutation/), to compare against.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import sklearn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "BRCA_DMS = pd.read_csv('Findlay_et_al_BRCA_DMS.txt', sep = '\\t', skiprows = 2)\n",
    "BRCA_DMS = BRCA_DMS[BRCA_DMS['func.class'] != 'None']\n",
    "BRCA_EVmutation = pd.read_csv(\"BRCA1_EVmutation.tsv\", sep = \"\\t\")\n",
    "BRCA_DMS['aa_pos'] = BRCA_DMS['aa_pos'].astype(str)\n",
    "BRCA_EVmutation['pos'] = BRCA_EVmutation['pos'].astype(str)\n",
    "BRCA_DMS['function.score.mean'] = BRCA_DMS['function.score.mean'].astype(float)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aa_pos</th>\n",
       "      <th>aa_ref</th>\n",
       "      <th>aa_alt</th>\n",
       "      <th>function.score.mean</th>\n",
       "      <th>func.class</th>\n",
       "      <th>uniprot</th>\n",
       "      <th>mutant</th>\n",
       "      <th>pos</th>\n",
       "      <th>wt</th>\n",
       "      <th>subs</th>\n",
       "      <th>prediction_epistatic</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>21</td>\n",
       "      <td>I</td>\n",
       "      <td>L</td>\n",
       "      <td>0.244698</td>\n",
       "      <td>0</td>\n",
       "      <td>P38398</td>\n",
       "      <td>I21L</td>\n",
       "      <td>21</td>\n",
       "      <td>I</td>\n",
       "      <td>L</td>\n",
       "      <td>-0.9572</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21</td>\n",
       "      <td>I</td>\n",
       "      <td>V</td>\n",
       "      <td>-0.147346</td>\n",
       "      <td>0</td>\n",
       "      <td>P38398</td>\n",
       "      <td>I21V</td>\n",
       "      <td>21</td>\n",
       "      <td>I</td>\n",
       "      <td>V</td>\n",
       "      <td>-0.9709</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>21</td>\n",
       "      <td>I</td>\n",
       "      <td>F</td>\n",
       "      <td>-0.432699</td>\n",
       "      <td>0</td>\n",
       "      <td>P38398</td>\n",
       "      <td>I21F</td>\n",
       "      <td>21</td>\n",
       "      <td>I</td>\n",
       "      <td>F</td>\n",
       "      <td>-2.5595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>21</td>\n",
       "      <td>I</td>\n",
       "      <td>N</td>\n",
       "      <td>-0.215463</td>\n",
       "      <td>0</td>\n",
       "      <td>P38398</td>\n",
       "      <td>I21N</td>\n",
       "      <td>21</td>\n",
       "      <td>I</td>\n",
       "      <td>N</td>\n",
       "      <td>1.5518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>21</td>\n",
       "      <td>I</td>\n",
       "      <td>T</td>\n",
       "      <td>0.072621</td>\n",
       "      <td>0</td>\n",
       "      <td>P38398</td>\n",
       "      <td>I21T</td>\n",
       "      <td>21</td>\n",
       "      <td>I</td>\n",
       "      <td>T</td>\n",
       "      <td>-0.0365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>688</th>\n",
       "      <td>1731</td>\n",
       "      <td>E</td>\n",
       "      <td>A</td>\n",
       "      <td>0.275906</td>\n",
       "      <td>0</td>\n",
       "      <td>P38398</td>\n",
       "      <td>E1731A</td>\n",
       "      <td>1731</td>\n",
       "      <td>E</td>\n",
       "      <td>A</td>\n",
       "      <td>-5.8658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>689</th>\n",
       "      <td>1731</td>\n",
       "      <td>E</td>\n",
       "      <td>G</td>\n",
       "      <td>0.065227</td>\n",
       "      <td>0</td>\n",
       "      <td>P38398</td>\n",
       "      <td>E1731G</td>\n",
       "      <td>1731</td>\n",
       "      <td>E</td>\n",
       "      <td>G</td>\n",
       "      <td>-8.0551</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>690</th>\n",
       "      <td>1731</td>\n",
       "      <td>E</td>\n",
       "      <td>V</td>\n",
       "      <td>-0.203848</td>\n",
       "      <td>0</td>\n",
       "      <td>P38398</td>\n",
       "      <td>E1731V</td>\n",
       "      <td>1731</td>\n",
       "      <td>E</td>\n",
       "      <td>V</td>\n",
       "      <td>-8.0551</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>691</th>\n",
       "      <td>1731</td>\n",
       "      <td>E</td>\n",
       "      <td>D</td>\n",
       "      <td>0.125433</td>\n",
       "      <td>0</td>\n",
       "      <td>P38398</td>\n",
       "      <td>E1731D</td>\n",
       "      <td>1731</td>\n",
       "      <td>E</td>\n",
       "      <td>D</td>\n",
       "      <td>-6.8724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>692</th>\n",
       "      <td>1731</td>\n",
       "      <td>E</td>\n",
       "      <td>D</td>\n",
       "      <td>-0.031138</td>\n",
       "      <td>0</td>\n",
       "      <td>P38398</td>\n",
       "      <td>E1731D</td>\n",
       "      <td>1731</td>\n",
       "      <td>E</td>\n",
       "      <td>D</td>\n",
       "      <td>-6.8724</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>693 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    aa_pos aa_ref aa_alt  function.score.mean func.class uniprot  mutant  \\\n",
       "0       21      I      L             0.244698          0  P38398    I21L   \n",
       "1       21      I      V            -0.147346          0  P38398    I21V   \n",
       "2       21      I      F            -0.432699          0  P38398    I21F   \n",
       "3       21      I      N            -0.215463          0  P38398    I21N   \n",
       "4       21      I      T             0.072621          0  P38398    I21T   \n",
       "..     ...    ...    ...                  ...        ...     ...     ...   \n",
       "688   1731      E      A             0.275906          0  P38398  E1731A   \n",
       "689   1731      E      G             0.065227          0  P38398  E1731G   \n",
       "690   1731      E      V            -0.203848          0  P38398  E1731V   \n",
       "691   1731      E      D             0.125433          0  P38398  E1731D   \n",
       "692   1731      E      D            -0.031138          0  P38398  E1731D   \n",
       "\n",
       "      pos wt subs  prediction_epistatic  \n",
       "0      21  I    L               -0.9572  \n",
       "1      21  I    V               -0.9709  \n",
       "2      21  I    F               -2.5595  \n",
       "3      21  I    N                1.5518  \n",
       "4      21  I    T               -0.0365  \n",
       "..    ... ..  ...                   ...  \n",
       "688  1731  E    A               -5.8658  \n",
       "689  1731  E    G               -8.0551  \n",
       "690  1731  E    V               -8.0551  \n",
       "691  1731  E    D               -6.8724  \n",
       "692  1731  E    D               -6.8724  \n",
       "\n",
       "[693 rows x 11 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "BRCA_DMS = pd.merge( BRCA_DMS, BRCA_EVmutation.drop_duplicates(), \\\n",
    "                     left_on = [\"aa_pos\", \"aa_ref\", \"aa_alt\"], \\\n",
    "                     right_on = [\"pos\", \"wt\", \"subs\"])\n",
    "BRCA_DMS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Obtaining variant annotation\n",
    "\n",
    "The first step is to annotate the variants such that we have the features that are needed for the predictors to work. Specifically, we need the following features:\n",
    "\n",
    "1. **DNA trinucleotide mutational signature**: The [CDSMutSig](https://github.com/josef0731/CDSMutSig) R package provides functionalities to obtain this.\n",
    "2. **DNA Conservation scores** of the trinucleotide signature: From (1), we can simply query these scores (using packages such as `genomicScores` in R/Bioconductor) with the genomic coordinates mapped.\n",
    "3. **Solvent accessibility** information from mapping to structures: We use mappings provided in the [ZoomVar](https://fraternalilab.kcl.ac.uk/ZoomVar/) database. Mapping per amino acid position to structures from ZoomVar is included as a flat-file data table in the [Zenodo repository of this manuscript](https://doi.org/10.5281/zenodo.4726168).\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. DNA Mutational signature\n",
    "\n",
    "the following code snippet in R shows how to use CDSMutSig to obtain mappings for given protein(s) - it gives all missense variants which can be obtained by substituting one nucleotide, at any positions along the protein sequence:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENSP00000350283 ...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Fetching CDS for 1 proteins ... \n",
      "R[write to console]: 1 found\n",
      "\n",
      "R[write to console]: Checking CDS and protein sequence lengths ... \n",
      "R[write to console]: 1/1 OK\n",
      "\n",
      "R[write to console]: Fetching CDS for 1 proteins ... \n",
      "R[write to console]: 1 found\n",
      "\n",
      "R[write to console]: Checking CDS and protein sequence lengths ... \n",
      "R[write to console]: 1/1 OK\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%R -o mutsigs\n",
    "# Map genomic position, MutSig and phyloP scores\n",
    "\n",
    "ensp <- data.frame(\n",
    "  uniprot_id = c(\"P38398\"),\n",
    "  protein_id = c(\"ENSP00000350283\"), # manually checked by searching UniProt & Ensembl \n",
    "  stringsAsFactors = FALSE\n",
    ")\n",
    "\n",
    "# Use CDSMutSig\n",
    "library(CDSMutSig)\n",
    "mutsigs <- lapply(ensp$protein_id, function(id){\n",
    "  cat(paste0(id, \" ...\\n\"))\n",
    "  codons <- mapCodon(id, EnsDb.Hsapiens.v86::EnsDb.Hsapiens.v86,\n",
    "                     BSgenome.Hsapiens.NCBI.GRCh38::Hsapiens, \n",
    "                     seqLevelStyle = \"NCBI\")\n",
    "  if( is.null( codons ) ) return(NULL)\n",
    "  codons_muts <- mapMut(codons, getPossibleMissenseSNVs())\n",
    "  codons_mutsigs <- mapMutSig(id, nrow(codons), codons_muts, \n",
    "                              EnsDb.Hsapiens.v86::EnsDb.Hsapiens.v86,\n",
    "                              BSgenome.Hsapiens.NCBI.GRCh38::Hsapiens, \n",
    "                              seqLevelStyle = \"NCBI\")\n",
    "  codons_mutsigs$ensp <- id  \n",
    "  codons_mutsigs\n",
    "})\n",
    "       \n",
    "mutsigs <- do.call(\"rbind\", mutsigs)\n",
    "mutsigs <- merge( ensp, mutsigs, by.x = \"protein_id\", by.y = \"ensp\" )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_id</th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>chr</th>\n",
       "      <th>g_pos</th>\n",
       "      <th>AA_pos</th>\n",
       "      <th>WT_AA</th>\n",
       "      <th>MUT_AA</th>\n",
       "      <th>WT_codon</th>\n",
       "      <th>MUT_codon</th>\n",
       "      <th>MutSig</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSP00000350283</td>\n",
       "      <td>P38398</td>\n",
       "      <td>17</td>\n",
       "      <td>43124018.0</td>\n",
       "      <td>1</td>\n",
       "      <td>M</td>\n",
       "      <td>T</td>\n",
       "      <td>ATG</td>\n",
       "      <td>ACG</td>\n",
       "      <td>A[T&gt;C]G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSP00000350283</td>\n",
       "      <td>P38398</td>\n",
       "      <td>17</td>\n",
       "      <td>43124017.0</td>\n",
       "      <td>1</td>\n",
       "      <td>M</td>\n",
       "      <td>L</td>\n",
       "      <td>ATG</td>\n",
       "      <td>TTG</td>\n",
       "      <td>A[T&gt;A]T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSP00000350283</td>\n",
       "      <td>P38398</td>\n",
       "      <td>17</td>\n",
       "      <td>43124018.0</td>\n",
       "      <td>1</td>\n",
       "      <td>M</td>\n",
       "      <td>K</td>\n",
       "      <td>ATG</td>\n",
       "      <td>AAG</td>\n",
       "      <td>A[T&gt;A]G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSP00000350283</td>\n",
       "      <td>P38398</td>\n",
       "      <td>17</td>\n",
       "      <td>43124018.0</td>\n",
       "      <td>1</td>\n",
       "      <td>M</td>\n",
       "      <td>R</td>\n",
       "      <td>ATG</td>\n",
       "      <td>AGG</td>\n",
       "      <td>A[T&gt;G]G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSP00000350283</td>\n",
       "      <td>P38398</td>\n",
       "      <td>17</td>\n",
       "      <td>43124017.0</td>\n",
       "      <td>1</td>\n",
       "      <td>M</td>\n",
       "      <td>L</td>\n",
       "      <td>ATG</td>\n",
       "      <td>CTG</td>\n",
       "      <td>A[T&gt;G]T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13271</th>\n",
       "      <td>ENSP00000350283</td>\n",
       "      <td>P38398</td>\n",
       "      <td>17</td>\n",
       "      <td>43045802.0</td>\n",
       "      <td>1863</td>\n",
       "      <td>Y</td>\n",
       "      <td>*</td>\n",
       "      <td>TAC</td>\n",
       "      <td>TAG</td>\n",
       "      <td>A[C&gt;G]T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13272</th>\n",
       "      <td>ENSP00000350283</td>\n",
       "      <td>P38398</td>\n",
       "      <td>17</td>\n",
       "      <td>43045802.0</td>\n",
       "      <td>1863</td>\n",
       "      <td>Y</td>\n",
       "      <td>*</td>\n",
       "      <td>TAC</td>\n",
       "      <td>TAA</td>\n",
       "      <td>A[C&gt;A]T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13273</th>\n",
       "      <td>ENSP00000350283</td>\n",
       "      <td>P38398</td>\n",
       "      <td>17</td>\n",
       "      <td>43045800.0</td>\n",
       "      <td>1863</td>\n",
       "      <td>Y</td>\n",
       "      <td>D</td>\n",
       "      <td>TAC</td>\n",
       "      <td>GAC</td>\n",
       "      <td>C[T&gt;G]A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13274</th>\n",
       "      <td>ENSP00000350283</td>\n",
       "      <td>P38398</td>\n",
       "      <td>17</td>\n",
       "      <td>43045801.0</td>\n",
       "      <td>1863</td>\n",
       "      <td>Y</td>\n",
       "      <td>F</td>\n",
       "      <td>TAC</td>\n",
       "      <td>TTC</td>\n",
       "      <td>G[T&gt;A]A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13275</th>\n",
       "      <td>ENSP00000350283</td>\n",
       "      <td>P38398</td>\n",
       "      <td>17</td>\n",
       "      <td>43045801.0</td>\n",
       "      <td>1863</td>\n",
       "      <td>Y</td>\n",
       "      <td>S</td>\n",
       "      <td>TAC</td>\n",
       "      <td>TCC</td>\n",
       "      <td>G[T&gt;G]A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13275 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            protein_id uniprot_id chr       g_pos  AA_pos WT_AA MUT_AA  \\\n",
       "1      ENSP00000350283     P38398  17  43124018.0       1     M      T   \n",
       "2      ENSP00000350283     P38398  17  43124017.0       1     M      L   \n",
       "3      ENSP00000350283     P38398  17  43124018.0       1     M      K   \n",
       "4      ENSP00000350283     P38398  17  43124018.0       1     M      R   \n",
       "5      ENSP00000350283     P38398  17  43124017.0       1     M      L   \n",
       "...                ...        ...  ..         ...     ...   ...    ...   \n",
       "13271  ENSP00000350283     P38398  17  43045802.0    1863     Y      *   \n",
       "13272  ENSP00000350283     P38398  17  43045802.0    1863     Y      *   \n",
       "13273  ENSP00000350283     P38398  17  43045800.0    1863     Y      D   \n",
       "13274  ENSP00000350283     P38398  17  43045801.0    1863     Y      F   \n",
       "13275  ENSP00000350283     P38398  17  43045801.0    1863     Y      S   \n",
       "\n",
       "      WT_codon MUT_codon   MutSig  \n",
       "1          ATG       ACG  A[T>C]G  \n",
       "2          ATG       TTG  A[T>A]T  \n",
       "3          ATG       AAG  A[T>A]G  \n",
       "4          ATG       AGG  A[T>G]G  \n",
       "5          ATG       CTG  A[T>G]T  \n",
       "...        ...       ...      ...  \n",
       "13271      TAC       TAG  A[C>G]T  \n",
       "13272      TAC       TAA  A[C>A]T  \n",
       "13273      TAC       GAC  C[T>G]A  \n",
       "13274      TAC       TTC  G[T>A]A  \n",
       "13275      TAC       TCC  G[T>G]A  \n",
       "\n",
       "[13275 rows x 10 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutsigs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As shown above, this gives a full table of all possible missense variants, their genomic positions, and the corresponding trinucleotide signature (column `MutSig`).\n",
    "\n",
    "### 2. Conservation\n",
    "\n",
    "With the `chr` and `g_pos` columns above we can fetch the conservation scores of the trinucleotide motif surrounding the substitution. Here we use the genomicScores R package to fetch the phyloP score of the trinucleotide positions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/josefng/.local/lib/python3.8/site-packages/rpy2/robjects/pandas2ri.py:59: UserWarning: Error while trying to convert the column \"protein_id\". Fall back to string conversion. The error is: module 'pandas' has no attribute 'NA'\n",
      "  warnings.warn('Error while trying to convert '\n",
      "/home/josefng/.local/lib/python3.8/site-packages/rpy2/robjects/pandas2ri.py:59: UserWarning: Error while trying to convert the column \"uniprot_id\". Fall back to string conversion. The error is: module 'pandas' has no attribute 'NA'\n",
      "  warnings.warn('Error while trying to convert '\n",
      "/home/josefng/.local/lib/python3.8/site-packages/rpy2/robjects/pandas2ri.py:59: UserWarning: Error while trying to convert the column \"chr\". Fall back to string conversion. The error is: module 'pandas' has no attribute 'NA'\n",
      "  warnings.warn('Error while trying to convert '\n",
      "/home/josefng/.local/lib/python3.8/site-packages/rpy2/robjects/pandas2ri.py:59: UserWarning: Error while trying to convert the column \"WT_AA\". Fall back to string conversion. The error is: module 'pandas' has no attribute 'NA'\n",
      "  warnings.warn('Error while trying to convert '\n",
      "/home/josefng/.local/lib/python3.8/site-packages/rpy2/robjects/pandas2ri.py:59: UserWarning: Error while trying to convert the column \"MUT_AA\". Fall back to string conversion. The error is: module 'pandas' has no attribute 'NA'\n",
      "  warnings.warn('Error while trying to convert '\n",
      "/home/josefng/.local/lib/python3.8/site-packages/rpy2/robjects/pandas2ri.py:59: UserWarning: Error while trying to convert the column \"WT_codon\". Fall back to string conversion. The error is: module 'pandas' has no attribute 'NA'\n",
      "  warnings.warn('Error while trying to convert '\n",
      "/home/josefng/.local/lib/python3.8/site-packages/rpy2/robjects/pandas2ri.py:59: UserWarning: Error while trying to convert the column \"MUT_codon\". Fall back to string conversion. The error is: module 'pandas' has no attribute 'NA'\n",
      "  warnings.warn('Error while trying to convert '\n",
      "/home/josefng/.local/lib/python3.8/site-packages/rpy2/robjects/pandas2ri.py:59: UserWarning: Error while trying to convert the column \"MutSig\". Fall back to string conversion. The error is: module 'pandas' has no attribute 'NA'\n",
      "  warnings.warn('Error while trying to convert '\n",
      "R[write to console]: Loading required package: stats4\n",
      "\n",
      "R[write to console]: Loading required package: BiocGenerics\n",
      "\n",
      "R[write to console]: Loading required package: parallel\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: 'BiocGenerics'\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from 'package:parallel':\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from 'package:stats':\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from 'package:base':\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which, which.max, which.min\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: S4Vectors\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: 'S4Vectors'\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from 'package:base':\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: IRanges\n",
      "\n",
      "R[write to console]: Loading required package: GenomeInfoDb\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: 'GenomicScores'\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from 'package:utils':\n",
      "\n",
      "    citation\n",
      "\n",
      "\n",
      "R[write to console]: snapshotDate(): 2019-10-29\n",
      "\n",
      "R[write to console]: loading from cache\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%R -i mutsigs -o PhyloP\n",
    "\n",
    "# Use PhyloP                                  \n",
    "library(GenomicRanges)\n",
    "library(GenomicScores)\n",
    "library(IRanges)\n",
    "library(stringr)\n",
    "\n",
    "phast <- getGScores(\"phyloP100way.UCSC.hg38\")\n",
    "\n",
    "PhyloP <- sapply(-1:1, function(x){\n",
    "  gscores_phylop <- gscores(phast,\n",
    "        GenomicRanges::GRanges(seqnames = paste0(\"chr\", mutsigs[, \"chr\"]),\n",
    "                               ranges = IRanges::IRanges(\n",
    "                               start = mutsigs[, \"g_pos\"] + x,\n",
    "                               end = mutsigs[, \"g_pos\"] + x)))\n",
    "  gscores_phylop <- as.data.frame(gscores_phylop)\n",
    "  return( gscores_phylop[, 6] )\n",
    "})\n",
    "\n",
    "colnames( PhyloP ) <- paste(\"PhyloP\", -1:1, sep = \"_\")\n",
    "PhyloP <- as.data.frame(PhyloP)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_id</th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>chr</th>\n",
       "      <th>g_pos</th>\n",
       "      <th>AA_pos</th>\n",
       "      <th>WT_AA</th>\n",
       "      <th>MUT_AA</th>\n",
       "      <th>WT_codon</th>\n",
       "      <th>MUT_codon</th>\n",
       "      <th>MutSig</th>\n",
       "      <th>PhyloP_-1</th>\n",
       "      <th>PhyloP_0</th>\n",
       "      <th>PhyloP_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSP00000350283</td>\n",
       "      <td>P38398</td>\n",
       "      <td>17</td>\n",
       "      <td>43124018.0</td>\n",
       "      <td>1</td>\n",
       "      <td>M</td>\n",
       "      <td>T</td>\n",
       "      <td>ATG</td>\n",
       "      <td>ACG</td>\n",
       "      <td>A[T&gt;C]G</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSP00000350283</td>\n",
       "      <td>P38398</td>\n",
       "      <td>17</td>\n",
       "      <td>43124017.0</td>\n",
       "      <td>1</td>\n",
       "      <td>M</td>\n",
       "      <td>L</td>\n",
       "      <td>ATG</td>\n",
       "      <td>TTG</td>\n",
       "      <td>A[T&gt;A]T</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSP00000350283</td>\n",
       "      <td>P38398</td>\n",
       "      <td>17</td>\n",
       "      <td>43124018.0</td>\n",
       "      <td>1</td>\n",
       "      <td>M</td>\n",
       "      <td>K</td>\n",
       "      <td>ATG</td>\n",
       "      <td>AAG</td>\n",
       "      <td>A[T&gt;A]G</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSP00000350283</td>\n",
       "      <td>P38398</td>\n",
       "      <td>17</td>\n",
       "      <td>43124018.0</td>\n",
       "      <td>1</td>\n",
       "      <td>M</td>\n",
       "      <td>R</td>\n",
       "      <td>ATG</td>\n",
       "      <td>AGG</td>\n",
       "      <td>A[T&gt;G]G</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSP00000350283</td>\n",
       "      <td>P38398</td>\n",
       "      <td>17</td>\n",
       "      <td>43124017.0</td>\n",
       "      <td>1</td>\n",
       "      <td>M</td>\n",
       "      <td>L</td>\n",
       "      <td>ATG</td>\n",
       "      <td>CTG</td>\n",
       "      <td>A[T&gt;G]T</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13271</th>\n",
       "      <td>ENSP00000350283</td>\n",
       "      <td>P38398</td>\n",
       "      <td>17</td>\n",
       "      <td>43045802.0</td>\n",
       "      <td>1863</td>\n",
       "      <td>Y</td>\n",
       "      <td>*</td>\n",
       "      <td>TAC</td>\n",
       "      <td>TAG</td>\n",
       "      <td>A[C&gt;G]T</td>\n",
       "      <td>-0.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13272</th>\n",
       "      <td>ENSP00000350283</td>\n",
       "      <td>P38398</td>\n",
       "      <td>17</td>\n",
       "      <td>43045802.0</td>\n",
       "      <td>1863</td>\n",
       "      <td>Y</td>\n",
       "      <td>*</td>\n",
       "      <td>TAC</td>\n",
       "      <td>TAA</td>\n",
       "      <td>A[C&gt;A]T</td>\n",
       "      <td>-0.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13273</th>\n",
       "      <td>ENSP00000350283</td>\n",
       "      <td>P38398</td>\n",
       "      <td>17</td>\n",
       "      <td>43045800.0</td>\n",
       "      <td>1863</td>\n",
       "      <td>Y</td>\n",
       "      <td>D</td>\n",
       "      <td>TAC</td>\n",
       "      <td>GAC</td>\n",
       "      <td>C[T&gt;G]A</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>-0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13274</th>\n",
       "      <td>ENSP00000350283</td>\n",
       "      <td>P38398</td>\n",
       "      <td>17</td>\n",
       "      <td>43045801.0</td>\n",
       "      <td>1863</td>\n",
       "      <td>Y</td>\n",
       "      <td>F</td>\n",
       "      <td>TAC</td>\n",
       "      <td>TTC</td>\n",
       "      <td>G[T&gt;A]A</td>\n",
       "      <td>3.0</td>\n",
       "      <td>-0.5</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13275</th>\n",
       "      <td>ENSP00000350283</td>\n",
       "      <td>P38398</td>\n",
       "      <td>17</td>\n",
       "      <td>43045801.0</td>\n",
       "      <td>1863</td>\n",
       "      <td>Y</td>\n",
       "      <td>S</td>\n",
       "      <td>TAC</td>\n",
       "      <td>TCC</td>\n",
       "      <td>G[T&gt;G]A</td>\n",
       "      <td>3.0</td>\n",
       "      <td>-0.5</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13275 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            protein_id uniprot_id chr       g_pos AA_pos WT_AA MUT_AA  \\\n",
       "1      ENSP00000350283     P38398  17  43124018.0      1     M      T   \n",
       "2      ENSP00000350283     P38398  17  43124017.0      1     M      L   \n",
       "3      ENSP00000350283     P38398  17  43124018.0      1     M      K   \n",
       "4      ENSP00000350283     P38398  17  43124018.0      1     M      R   \n",
       "5      ENSP00000350283     P38398  17  43124017.0      1     M      L   \n",
       "...                ...        ...  ..         ...    ...   ...    ...   \n",
       "13271  ENSP00000350283     P38398  17  43045802.0   1863     Y      *   \n",
       "13272  ENSP00000350283     P38398  17  43045802.0   1863     Y      *   \n",
       "13273  ENSP00000350283     P38398  17  43045800.0   1863     Y      D   \n",
       "13274  ENSP00000350283     P38398  17  43045801.0   1863     Y      F   \n",
       "13275  ENSP00000350283     P38398  17  43045801.0   1863     Y      S   \n",
       "\n",
       "      WT_codon MUT_codon   MutSig  PhyloP_-1  PhyloP_0  PhyloP_1  \n",
       "1          ATG       ACG  A[T>C]G        4.0       4.0       0.5  \n",
       "2          ATG       TTG  A[T>A]T        4.0       4.0       4.0  \n",
       "3          ATG       AAG  A[T>A]G        4.0       4.0       0.5  \n",
       "4          ATG       AGG  A[T>G]G        4.0       4.0       0.5  \n",
       "5          ATG       CTG  A[T>G]T        4.0       4.0       4.0  \n",
       "...        ...       ...      ...        ...       ...       ...  \n",
       "13271      TAC       TAG  A[C>G]T       -0.5       0.0       4.0  \n",
       "13272      TAC       TAA  A[C>A]T       -0.5       0.0       4.0  \n",
       "13273      TAC       GAC  C[T>G]A        4.0       3.0      -0.5  \n",
       "13274      TAC       TTC  G[T>A]A        3.0      -0.5       0.0  \n",
       "13275      TAC       TCC  G[T>G]A        3.0      -0.5       0.0  \n",
       "\n",
       "[13275 rows x 13 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutsigs = pd.concat([mutsigs, PhyloP], axis = 1)\n",
    "mutsigs['AA_pos'] = mutsigs['AA_pos'].astype(str)\n",
    "mutsigs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The columns `PhyloP_-1`, `PhyloP_0`, `PhyloP_1` refer to the phyloP conservation scores at position -1 (i.e. 1 base 5' to the DNA substitution), 0 (i.e. the substituted DNA position) and +1 (i.e. 1 base 3' to the DNA substitution), for each missense variant annotated using CDSMutSig.\n",
    "\n",
    "### 3. Solvent accessibility\n",
    "\n",
    "We use mappings provided in the ZoomVar database, which in turns use the [POPS and POPSComp](https://github.com/Fraternalilab/POPScomp) programs to annotate the solvent accessibility of each residue in the protein structure. The full table containing all analysed human Uniprot entries can be found in the [Zenodo repository of this manuscript](https://doi.org/10.5281/zenodo.4726168)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>position</th>\n",
       "      <th>Q(SASA)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P38398</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P38398</td>\n",
       "      <td>2</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P38398</td>\n",
       "      <td>3</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P38398</td>\n",
       "      <td>4</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P38398</td>\n",
       "      <td>5</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1858</th>\n",
       "      <td>P38398</td>\n",
       "      <td>1859</td>\n",
       "      <td>0.9639</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1859</th>\n",
       "      <td>P38398</td>\n",
       "      <td>1860</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1860</th>\n",
       "      <td>P38398</td>\n",
       "      <td>1861</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1861</th>\n",
       "      <td>P38398</td>\n",
       "      <td>1862</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1862</th>\n",
       "      <td>P38398</td>\n",
       "      <td>1863</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1863 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     protein position Q(SASA)\n",
       "0     P38398        1    None\n",
       "1     P38398        2    None\n",
       "2     P38398        3    None\n",
       "3     P38398        4    None\n",
       "4     P38398        5    None\n",
       "...      ...      ...     ...\n",
       "1858  P38398     1859  0.9639\n",
       "1859  P38398     1860    None\n",
       "1860  P38398     1861    None\n",
       "1861  P38398     1862    None\n",
       "1862  P38398     1863    None\n",
       "\n",
       "[1863 rows x 3 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# the zoomvar Q(SASA) information\n",
    "zoomvar = pd.read_csv('BRCA1_ZoomVar.tsv', sep = \"\\t\", index_col=False)\n",
    "zoomvar = zoomvar[['protein', 'position', 'Q(SASA)']]\n",
    "zoomvar['position'] = zoomvar['position'].astype(str)\n",
    "zoomvar"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we merge the tables together. We will only retain those variants with all the required mappings, i.e. only variants mappable to structures & manifested as single nucleotide substitution."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aa_pos</th>\n",
       "      <th>aa_ref</th>\n",
       "      <th>aa_alt</th>\n",
       "      <th>function.score.mean</th>\n",
       "      <th>func.class</th>\n",
       "      <th>uniprot</th>\n",
       "      <th>mutant</th>\n",
       "      <th>pos</th>\n",
       "      <th>wt</th>\n",
       "      <th>subs</th>\n",
       "      <th>...</th>\n",
       "      <th>MUT_AA</th>\n",
       "      <th>WT_codon</th>\n",
       "      <th>MUT_codon</th>\n",
       "      <th>MutSig</th>\n",
       "      <th>PhyloP_-1</th>\n",
       "      <th>PhyloP_0</th>\n",
       "      <th>PhyloP_1</th>\n",
       "      <th>protein</th>\n",
       "      <th>position</th>\n",
       "      <th>Q(SASA)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>21</td>\n",
       "      <td>I</td>\n",
       "      <td>L</td>\n",
       "      <td>0.244698</td>\n",
       "      <td>0</td>\n",
       "      <td>P38398</td>\n",
       "      <td>I21L</td>\n",
       "      <td>21</td>\n",
       "      <td>I</td>\n",
       "      <td>L</td>\n",
       "      <td>...</td>\n",
       "      <td>L</td>\n",
       "      <td>ATC</td>\n",
       "      <td>CTC</td>\n",
       "      <td>A[T&gt;G]T</td>\n",
       "      <td>-3.0</td>\n",
       "      <td>-3.0</td>\n",
       "      <td>1.5</td>\n",
       "      <td>P38398</td>\n",
       "      <td>21</td>\n",
       "      <td>0.3845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21</td>\n",
       "      <td>I</td>\n",
       "      <td>V</td>\n",
       "      <td>-0.147346</td>\n",
       "      <td>0</td>\n",
       "      <td>P38398</td>\n",
       "      <td>I21V</td>\n",
       "      <td>21</td>\n",
       "      <td>I</td>\n",
       "      <td>V</td>\n",
       "      <td>...</td>\n",
       "      <td>V</td>\n",
       "      <td>ATC</td>\n",
       "      <td>GTC</td>\n",
       "      <td>A[T&gt;C]T</td>\n",
       "      <td>-3.0</td>\n",
       "      <td>-3.0</td>\n",
       "      <td>1.5</td>\n",
       "      <td>P38398</td>\n",
       "      <td>21</td>\n",
       "      <td>0.3845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>21</td>\n",
       "      <td>I</td>\n",
       "      <td>F</td>\n",
       "      <td>-0.432699</td>\n",
       "      <td>0</td>\n",
       "      <td>P38398</td>\n",
       "      <td>I21F</td>\n",
       "      <td>21</td>\n",
       "      <td>I</td>\n",
       "      <td>F</td>\n",
       "      <td>...</td>\n",
       "      <td>F</td>\n",
       "      <td>ATC</td>\n",
       "      <td>TTC</td>\n",
       "      <td>A[T&gt;A]T</td>\n",
       "      <td>-3.0</td>\n",
       "      <td>-3.0</td>\n",
       "      <td>1.5</td>\n",
       "      <td>P38398</td>\n",
       "      <td>21</td>\n",
       "      <td>0.3845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>21</td>\n",
       "      <td>I</td>\n",
       "      <td>N</td>\n",
       "      <td>-0.215463</td>\n",
       "      <td>0</td>\n",
       "      <td>P38398</td>\n",
       "      <td>I21N</td>\n",
       "      <td>21</td>\n",
       "      <td>I</td>\n",
       "      <td>N</td>\n",
       "      <td>...</td>\n",
       "      <td>N</td>\n",
       "      <td>ATC</td>\n",
       "      <td>AAC</td>\n",
       "      <td>A[T&gt;A]C</td>\n",
       "      <td>-3.0</td>\n",
       "      <td>1.5</td>\n",
       "      <td>2.0</td>\n",
       "      <td>P38398</td>\n",
       "      <td>21</td>\n",
       "      <td>0.3845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>21</td>\n",
       "      <td>I</td>\n",
       "      <td>T</td>\n",
       "      <td>0.072621</td>\n",
       "      <td>0</td>\n",
       "      <td>P38398</td>\n",
       "      <td>I21T</td>\n",
       "      <td>21</td>\n",
       "      <td>I</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>ATC</td>\n",
       "      <td>ACC</td>\n",
       "      <td>A[T&gt;C]C</td>\n",
       "      <td>-3.0</td>\n",
       "      <td>1.5</td>\n",
       "      <td>2.0</td>\n",
       "      <td>P38398</td>\n",
       "      <td>21</td>\n",
       "      <td>0.3845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>880</th>\n",
       "      <td>1731</td>\n",
       "      <td>E</td>\n",
       "      <td>V</td>\n",
       "      <td>-0.203848</td>\n",
       "      <td>0</td>\n",
       "      <td>P38398</td>\n",
       "      <td>E1731V</td>\n",
       "      <td>1731</td>\n",
       "      <td>E</td>\n",
       "      <td>V</td>\n",
       "      <td>...</td>\n",
       "      <td>V</td>\n",
       "      <td>GAG</td>\n",
       "      <td>GTG</td>\n",
       "      <td>C[T&gt;A]C</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>P38398</td>\n",
       "      <td>1731</td>\n",
       "      <td>0.1122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>881</th>\n",
       "      <td>1731</td>\n",
       "      <td>E</td>\n",
       "      <td>D</td>\n",
       "      <td>0.125433</td>\n",
       "      <td>0</td>\n",
       "      <td>P38398</td>\n",
       "      <td>E1731D</td>\n",
       "      <td>1731</td>\n",
       "      <td>E</td>\n",
       "      <td>D</td>\n",
       "      <td>...</td>\n",
       "      <td>D</td>\n",
       "      <td>GAG</td>\n",
       "      <td>GAT</td>\n",
       "      <td>C[C&gt;A]T</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>P38398</td>\n",
       "      <td>1731</td>\n",
       "      <td>0.1122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>882</th>\n",
       "      <td>1731</td>\n",
       "      <td>E</td>\n",
       "      <td>D</td>\n",
       "      <td>0.125433</td>\n",
       "      <td>0</td>\n",
       "      <td>P38398</td>\n",
       "      <td>E1731D</td>\n",
       "      <td>1731</td>\n",
       "      <td>E</td>\n",
       "      <td>D</td>\n",
       "      <td>...</td>\n",
       "      <td>D</td>\n",
       "      <td>GAG</td>\n",
       "      <td>GAC</td>\n",
       "      <td>C[C&gt;G]T</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>P38398</td>\n",
       "      <td>1731</td>\n",
       "      <td>0.1122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>883</th>\n",
       "      <td>1731</td>\n",
       "      <td>E</td>\n",
       "      <td>D</td>\n",
       "      <td>-0.031138</td>\n",
       "      <td>0</td>\n",
       "      <td>P38398</td>\n",
       "      <td>E1731D</td>\n",
       "      <td>1731</td>\n",
       "      <td>E</td>\n",
       "      <td>D</td>\n",
       "      <td>...</td>\n",
       "      <td>D</td>\n",
       "      <td>GAG</td>\n",
       "      <td>GAT</td>\n",
       "      <td>C[C&gt;A]T</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>P38398</td>\n",
       "      <td>1731</td>\n",
       "      <td>0.1122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>884</th>\n",
       "      <td>1731</td>\n",
       "      <td>E</td>\n",
       "      <td>D</td>\n",
       "      <td>-0.031138</td>\n",
       "      <td>0</td>\n",
       "      <td>P38398</td>\n",
       "      <td>E1731D</td>\n",
       "      <td>1731</td>\n",
       "      <td>E</td>\n",
       "      <td>D</td>\n",
       "      <td>...</td>\n",
       "      <td>D</td>\n",
       "      <td>GAG</td>\n",
       "      <td>GAC</td>\n",
       "      <td>C[C&gt;G]T</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>P38398</td>\n",
       "      <td>1731</td>\n",
       "      <td>0.1122</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>885 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    aa_pos aa_ref aa_alt  function.score.mean func.class uniprot  mutant  \\\n",
       "0       21      I      L             0.244698          0  P38398    I21L   \n",
       "1       21      I      V            -0.147346          0  P38398    I21V   \n",
       "2       21      I      F            -0.432699          0  P38398    I21F   \n",
       "3       21      I      N            -0.215463          0  P38398    I21N   \n",
       "4       21      I      T             0.072621          0  P38398    I21T   \n",
       "..     ...    ...    ...                  ...        ...     ...     ...   \n",
       "880   1731      E      V            -0.203848          0  P38398  E1731V   \n",
       "881   1731      E      D             0.125433          0  P38398  E1731D   \n",
       "882   1731      E      D             0.125433          0  P38398  E1731D   \n",
       "883   1731      E      D            -0.031138          0  P38398  E1731D   \n",
       "884   1731      E      D            -0.031138          0  P38398  E1731D   \n",
       "\n",
       "      pos wt subs  ...  MUT_AA WT_codon MUT_codon   MutSig  PhyloP_-1  \\\n",
       "0      21  I    L  ...       L      ATC       CTC  A[T>G]T       -3.0   \n",
       "1      21  I    V  ...       V      ATC       GTC  A[T>C]T       -3.0   \n",
       "2      21  I    F  ...       F      ATC       TTC  A[T>A]T       -3.0   \n",
       "3      21  I    N  ...       N      ATC       AAC  A[T>A]C       -3.0   \n",
       "4      21  I    T  ...       T      ATC       ACC  A[T>C]C       -3.0   \n",
       "..    ... ..  ...  ...     ...      ...       ...      ...        ...   \n",
       "880  1731  E    V  ...       V      GAG       GTG  C[T>A]C        1.0   \n",
       "881  1731  E    D  ...       D      GAG       GAT  C[C>A]T        5.0   \n",
       "882  1731  E    D  ...       D      GAG       GAC  C[C>G]T        5.0   \n",
       "883  1731  E    D  ...       D      GAG       GAT  C[C>A]T        5.0   \n",
       "884  1731  E    D  ...       D      GAG       GAC  C[C>G]T        5.0   \n",
       "\n",
       "    PhyloP_0 PhyloP_1 protein position Q(SASA)  \n",
       "0       -3.0      1.5  P38398       21  0.3845  \n",
       "1       -3.0      1.5  P38398       21  0.3845  \n",
       "2       -3.0      1.5  P38398       21  0.3845  \n",
       "3        1.5      2.0  P38398       21  0.3845  \n",
       "4        1.5      2.0  P38398       21  0.3845  \n",
       "..       ...      ...     ...      ...     ...  \n",
       "880      5.0      5.0  P38398     1731  0.1122  \n",
       "881      5.0      5.0  P38398     1731  0.1122  \n",
       "882      5.0      5.0  P38398     1731  0.1122  \n",
       "883      5.0      5.0  P38398     1731  0.1122  \n",
       "884      5.0      5.0  P38398     1731  0.1122  \n",
       "\n",
       "[885 rows x 27 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# merge tables\n",
    "BRCA_DMS = pd.merge( BRCA_DMS, mutsigs, \\\n",
    "                     left_on = [\"aa_pos\", \"aa_ref\", \"aa_alt\"], \\\n",
    "                     right_on = [\"AA_pos\", \"WT_AA\", \"MUT_AA\"])\n",
    "\n",
    "BRCA_DMS = pd.merge( BRCA_DMS, zoomvar, left_on = ['uniprot_id', 'AA_pos'], right_on = ['protein', 'position'])\n",
    "BRCA_DMS = BRCA_DMS[ BRCA_DMS['Q(SASA)'].str.find('None') == -1 ]\n",
    "\n",
    "BRCA_DMS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "BRCA_DMS.drop_duplicates().to_csv(\"BRCA1_DMS_annotated.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We now have all the features required for the predictors. Functions in  below process the features:\n",
    "\n",
    "1. parse DNA substitution type & base at position -1 and +1 separately into three features\n",
    "2. Normalise PhyloP score so they are in range [-1, 1]\n",
    "3. Scale Q(SASA) by subtracting 0.15. So that +ve value means surface (exposed) and -ve value means core (buried)\n",
    "4. One-hot encoding of the non-numeric features."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Data read in. Now process them\n",
      "PhyloP score normalised to [-1, 1].\n",
      "MutSig parsed.\n",
      "WT_AA one-hot encoding ...\n",
      "MUT_AA one-hot encoding ...\n",
      "-1 and +1 DNA pos one-hot encoding ...\n",
      "DNA substitution one-hot encoding ...\n"
     ]
    }
   ],
   "source": [
    "from helper import * # functions in helper.R in this repository\n",
    "\n",
    "X, Y = prepare_data('BRCA1_DMS_annotated.csv', delimiter=\",\", output_col = \"func.class\")\n",
    "# save the objects\n",
    "pickle.dump( (X, Y), open('BRCA1_DMS_annotated.pickle', 'wb') )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Apply the predictors\n",
    "\n",
    "Now we can apply the predictors. We have predictors trained on two datasets:\n",
    "\n",
    "1. `DMSexp`: trained on experimental DMS data covering 10 proteins.\n",
    "2. `EVmutation`: trained on computational DMS data given by a scoring system implemented in [EVmutation](https://marks.hms.harvard.edu/evmutation/).\n",
    "\n",
    "For each, we have trained predictors using these different combinations of features:\n",
    "\n",
    "* AA + Cons (i.e. Conservation) + MutSig + QSASA (i.e. Solvent accessibility)\n",
    "* AA + Cons + MutSig\n",
    "* AA + MutSig + QSASA\n",
    "* AA + MutSig\n",
    "* MutSig only\n",
    "* AA only\n",
    "\n",
    "Additionally, for `DMSexp` and `EVmutation`, we also trained a version of predictors where variants residing in the protein core were removed from training, to evaluate the importance of protein core positions to learning variant effects. We call these the *no-core* predictors.\n",
    "\n",
    "All predictors were trained using the Gradient Boosting Classifier method in scikit-learn, and can be loaded from the pickle objects included in the repository - see below for example.\n",
    "\n",
    "Here we apply the predictors on the BRCA1 variants and show the ROC-Area-under-curve (AUC) - a higher ROC-AUC signifies better prediction performance."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DMSexp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ROC-AUC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AA_Cons_MutSig_QSASA</th>\n",
       "      <td>0.668973</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AA_Cons_MutSig</th>\n",
       "      <td>0.606973</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AA_MutSig_QSASA</th>\n",
       "      <td>0.651754</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AA_MutSig</th>\n",
       "      <td>0.727465</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AA</th>\n",
       "      <td>0.767858</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MutSig</th>\n",
       "      <td>0.623852</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       ROC-AUC\n",
       "AA_Cons_MutSig_QSASA  0.668973\n",
       "AA_Cons_MutSig        0.606973\n",
       "AA_MutSig_QSASA       0.651754\n",
       "AA_MutSig             0.727465\n",
       "AA                    0.767858\n",
       "MutSig                0.623852"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "setups = {\"AA_Cons_MutSig_QSASA\": range(58), \"AA_Cons_MutSig\": range(57), \\\n",
    "          \"AA_MutSig_QSASA\": [i for i in range(54)] + [57], \"AA_MutSig\": range(54), \\\n",
    "          \"AA\": range(40), \"MutSig\": range(41, 54)}\n",
    "\n",
    "DMSexp = dict()\n",
    "\n",
    "for setup, col_id in setups.items():\n",
    "    with open('DMSexp/setup/' + setup + '/GBM.pickle', 'rb' ) as r:\n",
    "        model = pickle.load( r )\n",
    "        DMSexp[setup] = sklearn.metrics.roc_auc_score(Y, model.predict_proba( X[:, col_id] )[:, 1])\n",
    "\n",
    "pd.DataFrame.from_dict(DMSexp, orient='index', columns=['ROC-AUC'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### EVmutation\n",
    "\n",
    "The original EVmutation training set contains BRCA1 variants, therefore uses the version trained on variants excluding BRCA1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ROC-AUC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AA_Cons_MutSig_QSASA</th>\n",
       "      <td>0.849537</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AA_Cons_MutSig</th>\n",
       "      <td>0.775805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AA_MutSig_QSASA</th>\n",
       "      <td>0.847734</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AA_MutSig</th>\n",
       "      <td>0.760578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AA</th>\n",
       "      <td>0.768005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MutSig</th>\n",
       "      <td>0.555248</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       ROC-AUC\n",
       "AA_Cons_MutSig_QSASA  0.849537\n",
       "AA_Cons_MutSig        0.775805\n",
       "AA_MutSig_QSASA       0.847734\n",
       "AA_MutSig             0.760578\n",
       "AA                    0.768005\n",
       "MutSig                0.555248"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "EVmutation = dict()\n",
    "\n",
    "for setup, col_id in setups.items():\n",
    "    with open('EVmutation/setup/' + setup + '/GBM-excludeBRCA1-PTEN.pickle', 'rb' ) as r:\n",
    "        model = pickle.load( r )\n",
    "        EVmutation[setup] = sklearn.metrics.roc_auc_score(Y, model.predict_proba( X[:, col_id] )[:, 1])\n",
    "\n",
    "pd.DataFrame.from_dict(EVmutation, orient='index', columns=['ROC-AUC'])"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "767d51c1340bd893661ea55ea3124f6de3c7a262a8b4abca0554b478b1e2ff90"
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
