{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Gradient boosting trees trained on experimental DMS assays"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup\n",
    "\n",
    "* Data: DMS assay data from 9 datasets for all possible AA substitution generated by single nucleotide change (see `DMSexp_prepare_data.ipynb` for details on what has been done).\n",
    "* Feature vector: each variant is represented by 5 features; after one-hot encoding we have 20 + 20 + 4 + 6 + 4 + 3 + 1 = **58 features** - \n",
    "  + **WT Amino acid** (20 levels)\n",
    "  + **MUT Amino Acid** (20 levels)\n",
    "  + **trinucleotide context** (96 levels) (*separate one-hot encoding for positions -1 [4 features], 0 [6 features], +1 [4 features]*)\n",
    "  + phyloP **conservation score** [separately for positions -1, 0, 1] [*3 features*]\n",
    "  + **Q(SASA)** of the mutated position\n",
    "* Objective: Predict whether given variant is damaging (label '1') or neutral (label '0'). These were labelled in `DMSexp_prepare_data.ipynb`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import glob\n",
    "import re\n",
    "import random\n",
    "import pickle\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import sklearn\n",
    "import imblearn\n",
    "import time\n",
    "import shap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('data/MAVEdb/human_proteins_retained_ready.pickle', 'rb') as r:\n",
    "    X, Y, scoreset_id = pickle.load(r)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Idea\n",
    "\n",
    "### Dealing with Imbalanced labels\n",
    "\n",
    "An issue is the unequal number of variants labelled '1' and '0'. Therefore needs to take steps to ensure training/testing considers equal number of 1's and 0's - will need to oversample the smaller subgroup.\n",
    "\n",
    "Because each dataset has a variable number of 1's and 0's we need specific ways to address this other than \n",
    "simply by assigning different weights to 1 and 0 in the training step (ideally this weight needs to be adjusted for each separate DMS dataset.\n",
    "\n",
    "Therefore use the following scheme to establish training set - for each DMS dataset:\n",
    "\n",
    "1. Let $m$ be the total number of variants included. First set asside $0.2 * m$ variants as test set. Let $m' = 0.8 * m$, i.e. the remaining variants.\n",
    "\n",
    "\n",
    "2. For the remaining variants, let $(m'_0, m'_1)$ be the number of variants in this group which are labelled 0 and 1 respectively. Then sample, without replacement, separately from the 0 and 1 subsets $m'_s$ variants where $m'_s = min(m'_0, m'_1)$. \n",
    "\n",
    "Note for the smaller subgroup this means no sampling were effectively being done. The sampled subsets from the 0's and the 1's together form the training set.\n",
    "\n",
    "\n",
    "3. Repeat step (2) for 100 times to derive 100 training sets to train 100 models. The final prediction will be derived by majority voting (i.e. averaging) the 100 models.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.33009040880503143"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# just to see how imbalanced the labels are:\n",
    "# proportion of labels == 1 (damaging)\n",
    "np.mean(Y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sample_fixed_set(X_set, Y_set, exp_id_set, seed = 1, proportion = 0.2):\n",
    "    \"\"\"\n",
    "    Give a randomly sampled (the size given by proportion) set with identical distribution \n",
    "    separately from each DMS dataset.\n",
    "    Input:\n",
    "      - X_set: np.array of shape (n_samples, n_features) The feature matrix.\n",
    "      - Y_set: np.array of shape (n_samples, 1). The labels.\n",
    "      - exp_id_set: np.array of shape (n_samples, 1). The experiment ID for each observation.\n",
    "    Return:\n",
    "      a list of row IDs which satisfy the condition (i.e. proportion * m variants are retained\n",
    "      and m is the number of variants in each DMS dataset)\n",
    "    \"\"\"\n",
    "    exp_ids = list(set(exp_id_set.flatten()))\n",
    "    random.seed(seed) # set seed\n",
    "    n_samples = X_set.shape[0]\n",
    "    id_vec = range(n_samples)\n",
    "    id_keep = []  # row IDs of the obs to keep\n",
    "    for exp_id in exp_ids:\n",
    "        subset = [i for i in id_vec if exp_id_set[i] == exp_id]\n",
    "        m_s = int(np.floor(proportion * len(subset))) # number to be sampled from each subgroup\n",
    "        id_keep += random.sample(subset, m_s)\n",
    "    return id_keep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set aside 20% as test set.\n",
    "test_set_id = sample_fixed_set(X, Y, scoreset_id, proportion = 0.2, seed = 1234)\n",
    "train_set_id = [i for i in range(X.shape[0]) if i not in test_set_id]\n",
    "\n",
    "X_test = X[test_set_id, :]\n",
    "Y_test = Y[test_set_id, :]\n",
    "scoreset_id_test = scoreset_id[test_set_id, :]\n",
    "X_train = X[train_set_id, :]\n",
    "Y_train = Y[train_set_id, :]\n",
    "scoreset_id_train = scoreset_id[train_set_id, :]\n",
    "\n",
    "# save the whole train set\n",
    "with open('data/train_set.pickle', 'wb') as f:\n",
    "    pickle.dump((X_train, Y_train, scoreset_id_train), f)\n",
    "    \n",
    "# save the test set and forget about it for now\n",
    "with open('data/test_set.pickle', 'wb') as f:\n",
    "    pickle.dump((X_test, Y_test, scoreset_id_test), f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Architecture\n",
    "\n",
    "using GradientBoostingClassifier in scikit-learn. A standard hyperparameter grid search was carried out."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# force balanced dataset using SMOTE\n",
    "from imblearn.over_sampling import SMOTE\n",
    "X_train_resampled, Y_train_resampled = SMOTE().fit_resample(X_train, Y_train)\n",
    "\n",
    "with open('data/train_set_resampled.pickle', 'wb') as w:\n",
    "    pickle.dump((X_train_resampled, Y_train_resampled), w)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Best: 0.877941 using {'learning_rate': 0.1, 'max_depth': 6, 'n_estimators': 500, 'subsample': 0.7}\n"
     ]
    }
   ],
   "source": [
    "# ref https://machinelearningmastery.com/gradient-boosting-machine-ensemble-in-python/\n",
    "from sklearn.model_selection import RepeatedStratifiedKFold\n",
    "from sklearn.model_selection import GridSearchCV\n",
    "from sklearn.ensemble import GradientBoostingClassifier\n",
    "\n",
    "# define the model with default hyperparameters\n",
    "model = GradientBoostingClassifier()\n",
    "# define the grid of values to search\n",
    "grid = dict()\n",
    "grid['n_estimators'] = [50, 100, 500]\n",
    "grid['learning_rate'] = [0.01, 0.1, 1.0]\n",
    "grid['subsample'] = [0.5, 0.7, 1.0]\n",
    "grid['max_depth'] = [3, 4, 5, 6]\n",
    "\n",
    "# define the evaluation procedure\n",
    "cv = RepeatedStratifiedKFold(n_splits=10, n_repeats=3, random_state=1)\n",
    "\n",
    "# define the grid search procedure\n",
    "grid_search = GridSearchCV(estimator=model, param_grid=grid, n_jobs=-1, cv=cv, scoring='accuracy')\n",
    "\n",
    "# execute the grid search\n",
    "grid_result = grid_search.fit(X_train_resampled, Y_train_resampled)\n",
    "\n",
    "# summarize the best score and configuration\n",
    "print(\"Best: %f using %s\" % (grid_result.best_score_, grid_result.best_params_))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('Hyperparams-GBM.pickle', 'wb') as w:\n",
    "    pickle.dump( (grid_result.best_score_, grid_result.best_params_), w )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('data/train_set_resampled.pickle', 'rb') as r:\n",
    "    X_train_resampled, Y_train_resampled = pickle.load( r )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# now train the actual model\n",
    "model = GradientBoostingClassifier( learning_rate = 0.1, max_depth = 6, n_estimators = 500, subsample = 0.7 )\n",
    "\n",
    "model = model.fit(X_train_resampled, Y_train_resampled)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "mkdir setup\n",
    "mkdir setup/AA_Cons_MutSig_QSASA\n",
    "mkdir setup/AA_Cons_MutSig\n",
    "mkdir setup/AA_MutSig_QSASA\n",
    "mkdir setup/AA_MutSig\n",
    "mkdir setup/AA\n",
    "mkdir setup/MutSig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save the model\n",
    "with open('setup/AA_Cons_MutSig_QSASA/GBM.pickle', 'wb' ) as o:\n",
    "    pickle.dump( model, o )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### AA + MutSig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# use only the first 54 columns (ie WT [20] + MUT [20] amino acids + MutSig position -1 [4 bases], 0[6 combinations of SBS], 1 [4 bases])\n",
    "model = GradientBoostingClassifier( learning_rate = 0.1, max_depth = 6, n_estimators = 500, subsample = 0.7 )\n",
    "\n",
    "model = model.fit(X_train_resampled[:, 0:54], Y_train_resampled)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save the model\n",
    "with open('setup/AA_MutSig/GBM.pickle', 'wb' ) as o:\n",
    "    pickle.dump( model, o )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### AA + Cons + MutSig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# use only the first 54 columns (ie WT [20] + MUT [20] amino acids + MutSig position -1 [4 bases], 0[6 combinations of SBS], 1 [4 bases])\n",
    "model = GradientBoostingClassifier( learning_rate = 0.1, max_depth = 6, n_estimators = 500, subsample = 0.7 )\n",
    "\n",
    "model = model.fit(X_train_resampled[:, 0:57], Y_train_resampled)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save the model\n",
    "with open('setup/AA_Cons_MutSig/GBM.pickle', 'wb' ) as o:\n",
    "    pickle.dump( model, o )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### AA + MutSig + QSASA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# use only the first 54 columns + last column (QSASA)\n",
    "model = GradientBoostingClassifier( learning_rate = 0.1, max_depth = 6, n_estimators = 500, subsample = 0.7 )\n",
    "\n",
    "model = model.fit(X_train_resampled[:, [i for i in range(54)] + [57]], Y_train_resampled)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save the model\n",
    "with open('setup/AA_MutSig_QSASA/GBM.pickle', 'wb' ) as o:\n",
    "    pickle.dump( model, o )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### AA only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# use only the first 40 columns (ie WT [20] + MUT [20] amino acids\n",
    "model = GradientBoostingClassifier( learning_rate = 0.1, max_depth = 6, n_estimators = 500, subsample = 0.7 )\n",
    "\n",
    "model = model.fit(X_train_resampled[:, 0:40], Y_train_resampled)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save the model\n",
    "with open('setup/AA/GBM.pickle', 'wb' ) as o:\n",
    "    pickle.dump( model, o )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### MutSig only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "model = GradientBoostingClassifier( learning_rate = 0.1, max_depth = 6, n_estimators = 500, subsample = 0.7 )\n",
    "\n",
    "model = model.fit(X_train_resampled[:, 41:54], Y_train_resampled)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save the model\n",
    "with open('setup/MutSig/GBM.pickle', 'wb' ) as o:\n",
    "    pickle.dump( model, o )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Predictors without using core variants\n",
    "\n",
    "Just to see effect of removing the variants most likley to be in the mutational dark matter."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "surface = X_train_resampled[:, -1] >= 0.15"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove obs where Q(SASA) < 0.15 and retrain the model\n",
    "X_train_resampled = X_train_resampled[surface, :]\n",
    "Y_train_resampled = Y_train_resampled[surface]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.ensemble import GradientBoostingClassifier\n",
    "\n",
    "# AA Cons MutSig QSASA\n",
    "model = GradientBoostingClassifier( learning_rate = 0.1, max_depth = 6, n_estimators = 500, subsample = 0.7 )\n",
    "model = model.fit(X_train_resampled, Y_train_resampled)\n",
    "with open('setup/AA_Cons_MutSig_QSASA/GBM-no-core.pickle', 'wb' ) as o:\n",
    "    pickle.dump( model, o )\n",
    "    \n",
    "# AA_MutSig\n",
    "# use only the first 54 columns (ie WT [20] + MUT [20] amino acids + MutSig position -1 [4 bases], 0[6 combinations of SBS], 1 [4 bases])\n",
    "model = GradientBoostingClassifier( learning_rate = 0.1, max_depth = 6, n_estimators = 500, subsample = 0.7 )\n",
    "model = model.fit(X_train_resampled[:, 0:54], Y_train_resampled)\n",
    "with open('setup/AA_MutSig/GBM-no-core.pickle', 'wb' ) as o:\n",
    "    pickle.dump( model, o )\n",
    "    \n",
    "# AA Cons MutSig\n",
    "# use only the first 54 columns (ie WT [20] + MUT [20] amino acids + MutSig position -1 [4 bases], 0[6 combinations of SBS], 1 [4 bases])\n",
    "model = GradientBoostingClassifier( learning_rate = 0.1, max_depth = 6, n_estimators = 500, subsample = 0.7 )\n",
    "model = model.fit(X_train_resampled[:, 0:57], Y_train_resampled)\n",
    "with open('setup/AA_Cons_MutSig/GBM-no-core.pickle', 'wb' ) as o:\n",
    "    pickle.dump( model, o )\n",
    "    \n",
    "# AA MutSig QSASA    \n",
    "# use only the first 54 columns + last column (QSASA)\n",
    "model = GradientBoostingClassifier( learning_rate = 0.1, max_depth = 6, n_estimators = 500, subsample = 0.7 )\n",
    "model = model.fit(X_train_resampled[:, [i for i in range(54)] + [57]], Y_train_resampled)\n",
    "with open('setup/AA_MutSig_QSASA/GBM-no-core.pickle', 'wb' ) as o:\n",
    "    pickle.dump( model, o )\n",
    "    \n",
    "# AA only\n",
    "# use only the first 40 columns (ie WT [20] + MUT [20] amino acids\n",
    "model = GradientBoostingClassifier( learning_rate = 0.1, max_depth = 6, n_estimators = 500, subsample = 0.7 )\n",
    "model = model.fit(X_train_resampled[:, 0:40], Y_train_resampled)\n",
    "with open('setup/AA/GBM-no-core.pickle', 'wb' ) as o:\n",
    "    pickle.dump( model, o )\n",
    "    \n",
    "# MutSig Only\n",
    "model = GradientBoostingClassifier( learning_rate = 0.1, max_depth = 6, n_estimators = 500, subsample = 0.7 )\n",
    "model = model.fit(X_train_resampled[:, 41:54], Y_train_resampled)\n",
    "with open('setup/MutSig/GBM-no-core.pickle', 'wb' ) as o:\n",
    "    pickle.dump( model, o )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Evaluation\n",
    "\n",
    "## Holdout set from same proteins/experiments encounted in training\n",
    "\n",
    "These were the 20% variants removed as `(X_test, Y_test)`, i.e. coming from the same proteins/experiments used in training but withheld from model construction and training."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getROCcurve(model, X, Y, dataset_name, test_set, model_name, model_type):\n",
    "    \"\"\"\n",
    "    given the feature matrix X, run it through model and generate a pd.DataFrame of statistics needed for\n",
    "    plotting roc curve.\n",
    "    args:\n",
    "      model: sklearn model\n",
    "      X: feature matrix\n",
    "      Y: true labels\n",
    "      dataset_name: name of training dataset (EVmutation / DMSexp)\n",
    "      test_set: name of testing dataset ('same protein'/'BRCA1'/etc)\n",
    "      model_name: name of model (e.g. 'AA_Cons_MutSig_QSASA')\n",
    "      model_type: type of model (classical / no-core)\n",
    "    \"\"\"\n",
    "    roc = pd.DataFrame(list(sklearn.metrics.roc_curve( Y, \\\n",
    "                                                       model.predict_proba( X )[:, 1] ))).T\n",
    "    roc.columns = ['fpr', 'tpr', 'threshold']\n",
    "    roc['dataset_name'] = dataset_name\n",
    "    roc['test_set'] = test_set\n",
    "    roc['model_name'] = model_name\n",
    "    roc['model_type'] = model_type\n",
    "    return roc    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AA_Cons_MutSig_QSASA :: Accuracy =  0.825178264076715 , F1-score =  0.7138832997987928 , ROC-AUC =  0.8683141553091445\n",
      "AA_Cons_MutSig       :: Accuracy =  0.7917383820998278 , F1-score =  0.6492753623188406 , ROC-AUC =  0.8265829261562114\n",
      "AA_MutSig_QSASA      :: Accuracy =  0.7831325301204819 , F1-score =  0.6903089887640449 , ROC-AUC =  0.8370406801153747\n",
      "AA_MutSig            :: Accuracy =  0.6006884681583476 , F1-score =  0.5045759609517999 , ROC-AUC =  0.6452008491039183\n",
      "AA only              :: Accuracy =  0.5960167199409885 , F1-score =  0.500456065673457 , ROC-AUC =  0.6366123502241573\n",
      "MutSig only          :: Accuracy =  0.578067371526924 , F1-score =  0.4417696811971373 , ROC-AUC =  0.5875954612252765\n"
     ]
    }
   ],
   "source": [
    "with open('setup/AA_Cons_MutSig_QSASA/GBM.pickle', 'rb' ) as r:\n",
    "    model = pickle.load( r )\n",
    "with open('data/test_set.pickle', 'rb') as r:\n",
    "    X_test, Y_test, scoreset_test = pickle.load(r)\n",
    "print('AA_Cons_MutSig_QSASA :: Accuracy = ', model.score( X_test, Y_test ), \\\n",
    "      \", F1-score = \", sklearn.metrics.f1_score(Y_test, model.predict( X_test ), average='binary'), \\\n",
    "      \", ROC-AUC = \", sklearn.metrics.roc_auc_score(Y_test, model.predict_proba( X_test )[:, 1]))\n",
    "roc = getROCcurve( model, X_test, Y_test, 'DMSexp', 'same proteins', 'AA + Cons + MutSig + Q(SASA)', 'all variants')\n",
    "roc.to_csv('roc_curves_all.csv', header=True, index=False, mode = 'a')\n",
    "\n",
    "with open('setup/AA_Cons_MutSig/GBM.pickle', 'rb' ) as r:\n",
    "    model = pickle.load( r )\n",
    "print('AA_Cons_MutSig       :: Accuracy = ', model.score( X_test[:, 0:57], Y_test ), \\\n",
    "      \", F1-score = \", sklearn.metrics.f1_score(Y_test, model.predict( X_test[:, 0:57], ), average='binary'), \\\n",
    "      \", ROC-AUC = \", sklearn.metrics.roc_auc_score(Y_test, model.predict_proba( X_test[:, 0:57] )[:, 1]))\n",
    "roc = getROCcurve( model, X_test[:, 0:57], Y_test, 'DMSexp', 'same proteins', 'AA + Cons + MutSig', 'all variants')\n",
    "roc.to_csv('roc_curves_all.csv', header=True, index=False, mode = 'a')\n",
    "\n",
    "with open('setup/AA_MutSig_QSASA/GBM.pickle', 'rb' ) as r:\n",
    "    model = pickle.load( r )\n",
    "print('AA_MutSig_QSASA      :: Accuracy = ', model.score( X_test[:, [i for i in range(54)] + [57]], Y_test ), \\\n",
    "      \", F1-score = \", sklearn.metrics.f1_score(Y_test, model.predict( X_test[:, [i for i in range(54)] + [57]] ), average='binary'), \\\n",
    "      \", ROC-AUC = \", sklearn.metrics.roc_auc_score(Y_test, model.predict_proba( X_test[:, [i for i in range(54)] + [57]] )[:, 1]))\n",
    "roc = getROCcurve( model, X_test[:, [i for i in range(54)] + [57]], Y_test, 'DMSexp', 'same proteins', 'AA + MutSig + Q(SASA)', 'all variants')\n",
    "roc.to_csv('roc_curves_all.csv', header=True, index=False, mode = 'a')\n",
    "\n",
    "with open('setup/AA_MutSig/GBM.pickle', 'rb' ) as r:\n",
    "    model = pickle.load( r )\n",
    "print('AA_MutSig            :: Accuracy = ', model.score( X_test[:, 0:54], Y_test ), \\\n",
    "      \", F1-score = \", sklearn.metrics.f1_score(Y_test, model.predict( X_test[:, 0:54] ), average='binary'), \\\n",
    "      \", ROC-AUC = \", sklearn.metrics.roc_auc_score(Y_test, model.predict_proba( X_test[:, 0:54] )[:, 1]))\n",
    "roc = getROCcurve( model, X_test[:, 0:54], Y_test, 'DMSexp', 'same proteins', 'AA + MutSig', 'all variants')\n",
    "roc.to_csv('roc_curves_all.csv', header=True, index=False, mode = 'a')\n",
    "\n",
    "with open('setup/AA/GBM.pickle', 'rb' ) as r:\n",
    "    model = pickle.load( r )\n",
    "print('AA only              :: Accuracy = ', model.score( X_test[:, 0:40], Y_test ), \\\n",
    "      \", F1-score = \", sklearn.metrics.f1_score(Y_test, model.predict( X_test[:, 0:40] ), average='binary'), \\\n",
    "      \", ROC-AUC = \", sklearn.metrics.roc_auc_score(Y_test, model.predict_proba( X_test[:, 0:40] )[:, 1]))\n",
    "roc = getROCcurve( model, X_test[:, 0:40], Y_test, 'DMSexp', 'same proteins', 'AA only', 'all variants')\n",
    "roc.to_csv('roc_curves_all.csv', header=True, index=False, mode = 'a')\n",
    "\n",
    "with open('setup/MutSig/GBM.pickle', 'rb' ) as r:\n",
    "    model = pickle.load( r )\n",
    "print('MutSig only          :: Accuracy = ', model.score( X_test[:, 41:54], Y_test ), \\\n",
    "      \", F1-score = \", sklearn.metrics.f1_score(Y_test, model.predict( X_test[:, 41:54] ), average='binary'), \\\n",
    "      \", ROC-AUC = \", sklearn.metrics.roc_auc_score(Y_test, model.predict_proba( X_test[:, 41:54] )[:, 1]))\n",
    "roc = getROCcurve( model, X_test[:, 41:54], Y_test, 'DMSexp', 'same proteins', 'MutSig only', 'all variants')\n",
    "roc.to_csv('roc_curves_all.csv', header=True, index=False, mode = 'a')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "82.5% accuracy!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using the 'no-core' predictors:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AA_Cons_MutSig_QSASA :: Accuracy =  0.7115810179493484 , F1-score =  0.5666789804211304 , ROC-AUC =  0.7487479440384232\n",
      "AA_Cons_MutSig       :: Accuracy =  0.7241209736906811 , F1-score =  0.4595375722543353 , ROC-AUC =  0.6810337133896484\n",
      "AA_MutSig_QSASA      :: Accuracy =  0.6528153430046717 , F1-score =  0.5866510538641687 , ROC-AUC =  0.7168099086945\n",
      "AA_MutSig            :: Accuracy =  0.6188836980575363 , F1-score =  0.3541666666666667 , ROC-AUC =  0.5762793227755645\n",
      "AA only              :: Accuracy =  0.6284730759773789 , F1-score =  0.36592530423835506 , ROC-AUC =  0.5825257775422197\n",
      "MutSig only          :: Accuracy =  0.6437177280550774 , F1-score =  0.28160634605850277 , ROC-AUC =  0.5636135317860969\n"
     ]
    }
   ],
   "source": [
    "with open('setup/AA_Cons_MutSig_QSASA/GBM-no-core.pickle', 'rb' ) as r:\n",
    "    model = pickle.load( r )\n",
    "print('AA_Cons_MutSig_QSASA :: Accuracy = ', model.score( X_test, Y_test ), \\\n",
    "      \", F1-score = \", sklearn.metrics.f1_score(Y_test, model.predict( X_test ), average='binary'), \\\n",
    "      \", ROC-AUC = \", sklearn.metrics.roc_auc_score(Y_test, model.predict_proba( X_test )[:, 1]))\n",
    "roc = getROCcurve( model, X_test, Y_test, 'DMSexp', 'same proteins', 'AA + Cons + MutSig + Q(SASA)', 'no-core')\n",
    "roc.to_csv('roc_curves_all.csv', header=True, index=False, mode = 'a')\n",
    "\n",
    "with open('setup/AA_Cons_MutSig/GBM-no-core.pickle', 'rb' ) as r:\n",
    "    model = pickle.load( r )\n",
    "print('AA_Cons_MutSig       :: Accuracy = ', model.score( X_test[:, 0:57], Y_test ), \\\n",
    "      \", F1-score = \", sklearn.metrics.f1_score(Y_test, model.predict( X_test[:, 0:57], ), average='binary'), \\\n",
    "      \", ROC-AUC = \", sklearn.metrics.roc_auc_score(Y_test, model.predict_proba( X_test[:, 0:57] )[:, 1]))\n",
    "roc = getROCcurve( model, X_test[:, 0:57], Y_test, 'DMSexp', 'same proteins', 'AA + Cons + MutSig', 'no-core')\n",
    "roc.to_csv('roc_curves_all.csv', header=True, index=False, mode = 'a')\n",
    "\n",
    "with open('setup/AA_MutSig_QSASA/GBM-no-core.pickle', 'rb' ) as r:\n",
    "    model = pickle.load( r )\n",
    "print('AA_MutSig_QSASA      :: Accuracy = ', model.score( X_test[:, [i for i in range(54)] + [57]], Y_test ), \\\n",
    "      \", F1-score = \", sklearn.metrics.f1_score(Y_test, model.predict( X_test[:, [i for i in range(54)] + [57]] ), average='binary'), \\\n",
    "      \", ROC-AUC = \", sklearn.metrics.roc_auc_score(Y_test, model.predict_proba( X_test[:, [i for i in range(54)] + [57]] )[:, 1]))\n",
    "roc = getROCcurve( model, X_test[:, [i for i in range(54)] + [57]], Y_test, 'DMSexp', 'same proteins', 'AA + MutSig + Q(SASA)', 'no-core')\n",
    "roc.to_csv('roc_curves_all.csv', header=True, index=False, mode = 'a')\n",
    "\n",
    "with open('setup/AA_MutSig/GBM-no-core.pickle', 'rb' ) as r:\n",
    "    model = pickle.load( r )\n",
    "print('AA_MutSig            :: Accuracy = ', model.score( X_test[:, 0:54], Y_test ), \\\n",
    "      \", F1-score = \", sklearn.metrics.f1_score(Y_test, model.predict( X_test[:, 0:54] ), average='binary'), \\\n",
    "      \", ROC-AUC = \", sklearn.metrics.roc_auc_score(Y_test, model.predict_proba( X_test[:, 0:54] )[:, 1]))\n",
    "roc = getROCcurve( model, X_test[:, 0:54], Y_test, 'DMSexp', 'same proteins', 'AA + MutSig', 'no-core')\n",
    "roc.to_csv('roc_curves_all.csv', header=True, index=False, mode = 'a')\n",
    "\n",
    "with open('setup/AA/GBM-no-core.pickle', 'rb' ) as r:\n",
    "    model = pickle.load( r )\n",
    "print('AA only              :: Accuracy = ', model.score( X_test[:, 0:40], Y_test ), \\\n",
    "      \", F1-score = \", sklearn.metrics.f1_score(Y_test, model.predict( X_test[:, 0:40] ), average='binary'), \\\n",
    "      \", ROC-AUC = \", sklearn.metrics.roc_auc_score(Y_test, model.predict_proba( X_test[:, 0:40] )[:, 1]))\n",
    "roc = getROCcurve( model, X_test[:, 0:40], Y_test, 'DMSexp', 'same proteins', 'AA only', 'no-core')\n",
    "roc.to_csv('roc_curves_all.csv', header=True, index=False, mode = 'a')\n",
    "\n",
    "with open('setup/MutSig/GBM-no-core.pickle', 'rb' ) as r:\n",
    "    model = pickle.load( r )\n",
    "print('MutSig only          :: Accuracy = ', model.score( X_test[:, 41:54], Y_test ), \\\n",
    "      \", F1-score = \", sklearn.metrics.f1_score(Y_test, model.predict( X_test[:, 41:54] ), average='binary'), \\\n",
    "      \", ROC-AUC = \", sklearn.metrics.roc_auc_score(Y_test, model.predict_proba( X_test[:, 41:54] )[:, 1]))\n",
    "roc = getROCcurve( model, X_test[:, 41:54], Y_test, 'DMSexp', 'same proteins', 'MutSig only', 'no-core')\n",
    "roc.to_csv('roc_curves_all.csv', header=True, index=False, mode = 'a')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Independent benchmark\n",
    "\n",
    "1. Loss-of-function and Functional variants as annotated by BRCA1 DMS experiment ([Finlay et al Nature 2018](https://www.nature.com/articles/s41586-018-0461-z))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AA_Cons_MutSig_QSASA :: Accuracy =  0.7186440677966102 , F1-score =  0.3821339950372209 , ROC-AUC =  0.6689726605457661\n",
      "AA_Cons_MutSig       :: Accuracy =  0.6915254237288135 , F1-score =  0.20869565217391303\n",
      "AA_Cons_MutSig       :: Accuracy =  0.6915254237288135 , F1-score =  0.20869565217391303 , ROC-AUC =  0.6069729674432879\n",
      "AA_MutSig_QSASA      :: Accuracy =  0.6259887005649718 , F1-score =  0.4582651391162029 , ROC-AUC =  0.651754430832971\n",
      "AA_MutSig            :: Accuracy =  0.6971751412429379 , F1-score =  0.5281690140845071 , ROC-AUC =  0.7274654100918135\n",
      "AA only              :: Accuracy =  0.7310734463276836 , F1-score =  0.6020066889632106 , ROC-AUC =  0.7678575995498835\n",
      "MutSig only          :: Accuracy =  0.6474576271186441 , F1-score =  0.42857142857142855 , ROC-AUC =  0.6238523311424261\n",
      "\n",
      "no-core predictors:\n",
      "\n",
      "AA_Cons_MutSig_QSASA :: Accuracy =  0.6937853107344633 , F1-score =  0.31043256997455476 , ROC-AUC =  0.6902700698191863\n",
      "AA_Cons_MutSig       :: Accuracy =  0.672316384180791 , F1-score =  0.12121212121212119 , ROC-AUC =  0.5179854735173013\n",
      "AA_MutSig_QSASA      :: Accuracy =  0.5988700564971752 , F1-score =  0.5322793148880106 , ROC-AUC =  0.7287089844249508\n",
      "AA_MutSig            :: Accuracy =  0.6971751412429379 , F1-score =  0.40178571428571425 , ROC-AUC =  0.5747135623129843\n",
      "AA only              :: Accuracy =  0.6485875706214689 , F1-score =  0.3452631578947368 , ROC-AUC =  0.5884951791514283\n",
      "MutSig only          :: Accuracy =  0.7163841807909604 , F1-score =  0.25958702064896755 , ROC-AUC =  0.5767819237359659\n"
     ]
    }
   ],
   "source": [
    "with open('setup/AA_Cons_MutSig_QSASA/GBM.pickle', 'rb' ) as r:\n",
    "    model = pickle.load( r )\n",
    "with open('../EVmutation/human_protein_predictions/BRCA_DMS_annotated.pickle', 'rb') as r:\n",
    "    X_test, Y_test = pickle.load(r)\n",
    "print('AA_Cons_MutSig_QSASA :: Accuracy = ', model.score( X_test, Y_test ), \\\n",
    "      \", F1-score = \", sklearn.metrics.f1_score(Y_test, model.predict( X_test ), average='binary'), \\\n",
    "      \", ROC-AUC = \", sklearn.metrics.roc_auc_score(Y_test, model.predict_proba( X_test )[:, 1]))\n",
    "roc = getROCcurve( model, X_test, Y_test, 'DMSexp', 'BRCA1 DMS', 'AA + Cons + MutSig + Q(SASA)', 'all variants')\n",
    "roc.to_csv('roc_curves_all.csv', header=True, index=False, mode = 'a')\n",
    "\n",
    "with open('setup/AA_Cons_MutSig/GBM.pickle', 'rb' ) as r:\n",
    "    model = pickle.load( r )\n",
    "print('AA_Cons_MutSig       :: Accuracy = ', model.score( X_test[:, 0:57], Y_test ), \\\n",
    "      \", F1-score = \", sklearn.metrics.f1_score(Y_test, model.predict( X_test[:, 0:57], ), average='binary'), \\\n",
    "      \", ROC-AUC = \", sklearn.metrics.roc_auc_score(Y_test, model.predict_proba( X_test[:, 0:57] )[:, 1]))\n",
    "roc = getROCcurve( model, X_test[:, 0:57], Y_test, 'DMSexp', 'BRCA1 DMS', 'AA + Cons + MutSig', 'all variants')\n",
    "roc.to_csv('roc_curves_all.csv', header=True, index=False, mode = 'a')\n",
    "\n",
    "with open('setup/AA_MutSig_QSASA/GBM.pickle', 'rb' ) as r:\n",
    "    model = pickle.load( r )\n",
    "print('AA_MutSig_QSASA      :: Accuracy = ', model.score( X_test[:, [i for i in range(54)] + [57]], Y_test ), \\\n",
    "      \", F1-score = \", sklearn.metrics.f1_score(Y_test, model.predict( X_test[:, [i for i in range(54)] + [57]] ), average='binary'), \\\n",
    "      \", ROC-AUC = \", sklearn.metrics.roc_auc_score(Y_test, model.predict_proba( X_test[:, [i for i in range(54)] + [57]] )[:, 1]))\n",
    "roc = getROCcurve( model, X_test[:, [i for i in range(54)] + [57]], Y_test, 'DMSexp', 'BRCA1 DMS', \\\n",
    "                  'AA + MutSig + Q(SASA)', 'all variants')\n",
    "roc.to_csv('roc_curves_all.csv', header=True, index=False, mode = 'a')\n",
    "\n",
    "with open('setup/AA_MutSig/GBM.pickle', 'rb' ) as r:\n",
    "    model = pickle.load( r )\n",
    "print('AA_MutSig            :: Accuracy = ', model.score( X_test[:, 0:54], Y_test ), \\\n",
    "      \", F1-score = \", sklearn.metrics.f1_score(Y_test, model.predict( X_test[:, 0:54] ), average='binary'), \\\n",
    "      \", ROC-AUC = \", sklearn.metrics.roc_auc_score(Y_test, model.predict_proba( X_test[:, 0:54] )[:, 1]))\n",
    "roc = getROCcurve( model, X_test[:, 0:54], Y_test, 'DMSexp', 'BRCA1 DMS', 'AA + MutSig', 'all variants')\n",
    "roc.to_csv('roc_curves_all.csv', header=True, index=False, mode = 'a')\n",
    "\n",
    "with open('setup/AA/GBM.pickle', 'rb' ) as r:\n",
    "    model = pickle.load( r )\n",
    "print('AA only              :: Accuracy = ', model.score( X_test[:, 0:40], Y_test ), \\\n",
    "      \", F1-score = \", sklearn.metrics.f1_score(Y_test, model.predict( X_test[:, 0:40] ), average='binary'), \\\n",
    "      \", ROC-AUC = \", sklearn.metrics.roc_auc_score(Y_test, model.predict_proba( X_test[:, 0:40] )[:, 1]))\n",
    "roc = getROCcurve( model, X_test[:, 0:40], Y_test, 'DMSexp', 'BRCA1 DMS', 'AA only', 'all variants')\n",
    "roc.to_csv('roc_curves_all.csv', header=True, index=False, mode = 'a')\n",
    "\n",
    "with open('setup/MutSig/GBM.pickle', 'rb' ) as r:\n",
    "    model = pickle.load( r )\n",
    "print('MutSig only          :: Accuracy = ', model.score( X_test[:, 41:54], Y_test ), \\\n",
    "      \", F1-score = \", sklearn.metrics.f1_score(Y_test, model.predict( X_test[:, 41:54] ), average='binary'), \\\n",
    "      \", ROC-AUC = \", sklearn.metrics.roc_auc_score(Y_test, model.predict_proba( X_test[:, 41:54] )[:, 1]))\n",
    "roc = getROCcurve( model, X_test[:, 41:54], Y_test, 'DMSexp', 'BRCA1 DMS', 'MutSig only', 'all variants')\n",
    "roc.to_csv('roc_curves_all.csv', header=True, index=False, mode = 'a')\n",
    "\n",
    "# no-core\n",
    "print('\\nno-core predictors:\\n')\n",
    "with open('setup/AA_Cons_MutSig_QSASA/GBM-no-core.pickle', 'rb' ) as r:\n",
    "    model = pickle.load( r )\n",
    "print('AA_Cons_MutSig_QSASA :: Accuracy = ', model.score( X_test, Y_test ), \\\n",
    "      \", F1-score = \", sklearn.metrics.f1_score(Y_test, model.predict( X_test ), average='binary'), \\\n",
    "      \", ROC-AUC = \", sklearn.metrics.roc_auc_score(Y_test, model.predict_proba( X_test )[:, 1]))\n",
    "roc = getROCcurve( model, X_test, Y_test, 'DMSexp', 'BRCA1 DMS', 'AA + Cons + MutSig + Q(SASA)', 'no-core')\n",
    "roc.to_csv('roc_curves_all.csv', header=True, index=False, mode = 'a')\n",
    "\n",
    "with open('setup/AA_Cons_MutSig/GBM-no-core.pickle', 'rb' ) as r:\n",
    "    model = pickle.load( r )\n",
    "print('AA_Cons_MutSig       :: Accuracy = ', model.score( X_test[:, 0:57], Y_test ), \\\n",
    "      \", F1-score = \", sklearn.metrics.f1_score(Y_test, model.predict( X_test[:, 0:57], ), average='binary'), \\\n",
    "      \", ROC-AUC = \", sklearn.metrics.roc_auc_score(Y_test, model.predict_proba( X_test[:, 0:57] )[:, 1]))\n",
    "roc = getROCcurve( model, X_test[:, 0:57], Y_test, 'DMSexp', 'BRCA1 DMS', 'AA + Cons + MutSig', 'no-core')\n",
    "roc.to_csv('roc_curves_all.csv', header=True, index=False, mode = 'a')\n",
    "\n",
    "with open('setup/AA_MutSig_QSASA/GBM-no-core.pickle', 'rb' ) as r:\n",
    "    model = pickle.load( r )\n",
    "print('AA_MutSig_QSASA      :: Accuracy = ', model.score( X_test[:, [i for i in range(54)] + [57]], Y_test ), \\\n",
    "      \", F1-score = \", sklearn.metrics.f1_score(Y_test, model.predict( X_test[:, [i for i in range(54)] + [57]] ), average='binary'), \\\n",
    "      \", ROC-AUC = \", sklearn.metrics.roc_auc_score(Y_test, model.predict_proba( X_test[:, [i for i in range(54)] + [57]] )[:, 1]))\n",
    "roc = getROCcurve( model, X_test[:, [i for i in range(54)] + [57]], Y_test, 'DMSexp', 'BRCA1 DMS', 'AA + MutSig + Q(SASA)', 'no-core')\n",
    "roc.to_csv('roc_curves_all.csv', header=True, index=False, mode = 'a')\n",
    "\n",
    "with open('setup/AA_MutSig/GBM-no-core.pickle', 'rb' ) as r:\n",
    "    model = pickle.load( r )\n",
    "print('AA_MutSig            :: Accuracy = ', model.score( X_test[:, 0:54], Y_test ), \\\n",
    "      \", F1-score = \", sklearn.metrics.f1_score(Y_test, model.predict( X_test[:, 0:54] ), average='binary'), \\\n",
    "      \", ROC-AUC = \", sklearn.metrics.roc_auc_score(Y_test, model.predict_proba( X_test[:, 0:54] )[:, 1]))\n",
    "roc = getROCcurve( model, X_test[:, 0:54], Y_test, 'DMSexp', 'BRCA1 DMS', 'AA + MutSig', 'no-core')\n",
    "roc.to_csv('roc_curves_all.csv', header=True, index=False, mode = 'a')\n",
    "\n",
    "with open('setup/AA/GBM-no-core.pickle', 'rb' ) as r:\n",
    "    model = pickle.load( r )\n",
    "print('AA only              :: Accuracy = ', model.score( X_test[:, 0:40], Y_test ), \\\n",
    "      \", F1-score = \", sklearn.metrics.f1_score(Y_test, model.predict( X_test[:, 0:40] ), average='binary'), \\\n",
    "      \", ROC-AUC = \", sklearn.metrics.roc_auc_score(Y_test, model.predict_proba( X_test[:, 0:40] )[:, 1]))\n",
    "roc = getROCcurve( model, X_test[:, 0:40], Y_test, 'DMSexp', 'BRCA1 DMS', 'AA only', 'no-core')\n",
    "roc.to_csv('roc_curves_all.csv', header=True, index=False, mode = 'a')\n",
    "\n",
    "with open('setup/MutSig/GBM-no-core.pickle', 'rb' ) as r:\n",
    "    model = pickle.load( r )\n",
    "print('MutSig only          :: Accuracy = ', model.score( X_test[:, 41:54], Y_test ), \\\n",
    "      \", F1-score = \", sklearn.metrics.f1_score(Y_test, model.predict( X_test[:, 41:54] ), average='binary'), \\\n",
    "      \", ROC-AUC = \", sklearn.metrics.roc_auc_score(Y_test, model.predict_proba( X_test[:, 41:54] )[:, 1]))\n",
    "roc = getROCcurve( model, X_test[:, 41:54], Y_test, 'DMSexp', 'BRCA1 DMS', 'MutSig only', 'no-core')\n",
    "roc.to_csv('roc_curves_all.csv', header=True, index=False, mode = 'a')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2. Envision mutations\n",
    "\n",
    "Here only use the predict function (as Envision does not provide labels but rather a score. Will investigate correlation between Envision scores and our predicted damaging-probabiltiies."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
     ]
    }
   ],
   "source": [
    "with open('setup/AA_Cons_MutSig_QSASA/GBM.pickle', 'rb' ) as r:\n",
    "    model = pickle.load( r )\n",
    "with open('../Envision/Envision_ready_unique.pickle', 'rb') as r:\n",
    "    X_test, Y_test = pickle.load(r)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>X1</th>\n",
       "      <th>id2</th>\n",
       "      <th>AA1</th>\n",
       "      <th>AA2</th>\n",
       "      <th>position</th>\n",
       "      <th>Uniprot</th>\n",
       "      <th>WT_Mut</th>\n",
       "      <th>Variant</th>\n",
       "      <th>AA1_polarity</th>\n",
       "      <th>...</th>\n",
       "      <th>PhyloP_0</th>\n",
       "      <th>PhyloP_1</th>\n",
       "      <th>protein</th>\n",
       "      <th>Q(SASA)</th>\n",
       "      <th>AA_Cons_MutSig_QSASA_p</th>\n",
       "      <th>AA_Cons_MutSig_p</th>\n",
       "      <th>AA_MutSig_QSASA_p</th>\n",
       "      <th>AA_MutSig_p</th>\n",
       "      <th>AA_p</th>\n",
       "      <th>MutSig_p</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8323</td>\n",
       "      <td>19082162</td>\n",
       "      <td>P04035_P439A</td>\n",
       "      <td>P</td>\n",
       "      <td>A</td>\n",
       "      <td>439</td>\n",
       "      <td>P04035</td>\n",
       "      <td>PA</td>\n",
       "      <td>P439A</td>\n",
       "      <td>Special</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>P04035</td>\n",
       "      <td>1.0323</td>\n",
       "      <td>0.618426</td>\n",
       "      <td>0.195662</td>\n",
       "      <td>0.606245</td>\n",
       "      <td>0.541139</td>\n",
       "      <td>0.606757</td>\n",
       "      <td>0.408901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8329</td>\n",
       "      <td>19082168</td>\n",
       "      <td>P04035_P439H</td>\n",
       "      <td>P</td>\n",
       "      <td>H</td>\n",
       "      <td>439</td>\n",
       "      <td>P04035</td>\n",
       "      <td>PH</td>\n",
       "      <td>P439H</td>\n",
       "      <td>Special</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>-0.5</td>\n",
       "      <td>P04035</td>\n",
       "      <td>1.0323</td>\n",
       "      <td>0.098062</td>\n",
       "      <td>0.110471</td>\n",
       "      <td>0.592379</td>\n",
       "      <td>0.700575</td>\n",
       "      <td>0.650552</td>\n",
       "      <td>0.678710</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8332</td>\n",
       "      <td>19082171</td>\n",
       "      <td>P04035_P439L</td>\n",
       "      <td>P</td>\n",
       "      <td>L</td>\n",
       "      <td>439</td>\n",
       "      <td>P04035</td>\n",
       "      <td>PL</td>\n",
       "      <td>P439L</td>\n",
       "      <td>Special</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>-0.5</td>\n",
       "      <td>P04035</td>\n",
       "      <td>1.0323</td>\n",
       "      <td>0.226643</td>\n",
       "      <td>0.150819</td>\n",
       "      <td>0.728743</td>\n",
       "      <td>0.789619</td>\n",
       "      <td>0.671871</td>\n",
       "      <td>0.712546</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>8336</td>\n",
       "      <td>19082175</td>\n",
       "      <td>P04035_P439R</td>\n",
       "      <td>P</td>\n",
       "      <td>R</td>\n",
       "      <td>439</td>\n",
       "      <td>P04035</td>\n",
       "      <td>PR</td>\n",
       "      <td>P439R</td>\n",
       "      <td>Special</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>-0.5</td>\n",
       "      <td>P04035</td>\n",
       "      <td>1.0323</td>\n",
       "      <td>0.145746</td>\n",
       "      <td>0.185445</td>\n",
       "      <td>0.771501</td>\n",
       "      <td>0.782489</td>\n",
       "      <td>0.686362</td>\n",
       "      <td>0.685450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>8337</td>\n",
       "      <td>19082176</td>\n",
       "      <td>P04035_P439S</td>\n",
       "      <td>P</td>\n",
       "      <td>S</td>\n",
       "      <td>439</td>\n",
       "      <td>P04035</td>\n",
       "      <td>PS</td>\n",
       "      <td>P439S</td>\n",
       "      <td>Special</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>P04035</td>\n",
       "      <td>1.0323</td>\n",
       "      <td>0.387892</td>\n",
       "      <td>0.295273</td>\n",
       "      <td>0.441203</td>\n",
       "      <td>0.385010</td>\n",
       "      <td>0.555850</td>\n",
       "      <td>0.438892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8291</th>\n",
       "      <td>4610</td>\n",
       "      <td>3660589</td>\n",
       "      <td>Q9H3S4_S243N</td>\n",
       "      <td>S</td>\n",
       "      <td>N</td>\n",
       "      <td>243</td>\n",
       "      <td>Q9H3S4</td>\n",
       "      <td>SN</td>\n",
       "      <td>S243N</td>\n",
       "      <td>Polar</td>\n",
       "      <td>...</td>\n",
       "      <td>0.5</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Q9H3S4</td>\n",
       "      <td>1.2412</td>\n",
       "      <td>0.052334</td>\n",
       "      <td>0.085340</td>\n",
       "      <td>0.080482</td>\n",
       "      <td>0.420180</td>\n",
       "      <td>0.452736</td>\n",
       "      <td>0.297807</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8292</th>\n",
       "      <td>4613</td>\n",
       "      <td>3660592</td>\n",
       "      <td>Q9H3S4_S243R</td>\n",
       "      <td>S</td>\n",
       "      <td>R</td>\n",
       "      <td>243</td>\n",
       "      <td>Q9H3S4</td>\n",
       "      <td>SR</td>\n",
       "      <td>S243R</td>\n",
       "      <td>Polar</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>Q9H3S4</td>\n",
       "      <td>1.2412</td>\n",
       "      <td>0.349461</td>\n",
       "      <td>0.246390</td>\n",
       "      <td>0.278748</td>\n",
       "      <td>0.827931</td>\n",
       "      <td>0.586168</td>\n",
       "      <td>0.474273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8293</th>\n",
       "      <td>4613</td>\n",
       "      <td>3660592</td>\n",
       "      <td>Q9H3S4_S243R</td>\n",
       "      <td>S</td>\n",
       "      <td>R</td>\n",
       "      <td>243</td>\n",
       "      <td>Q9H3S4</td>\n",
       "      <td>SR</td>\n",
       "      <td>S243R</td>\n",
       "      <td>Polar</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>Q9H3S4</td>\n",
       "      <td>1.2412</td>\n",
       "      <td>0.248722</td>\n",
       "      <td>0.188052</td>\n",
       "      <td>0.158618</td>\n",
       "      <td>0.114509</td>\n",
       "      <td>0.586168</td>\n",
       "      <td>0.427768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8294</th>\n",
       "      <td>4613</td>\n",
       "      <td>3660592</td>\n",
       "      <td>Q9H3S4_S243R</td>\n",
       "      <td>S</td>\n",
       "      <td>R</td>\n",
       "      <td>243</td>\n",
       "      <td>Q9H3S4</td>\n",
       "      <td>SR</td>\n",
       "      <td>S243R</td>\n",
       "      <td>Polar</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>Q9H3S4</td>\n",
       "      <td>1.2412</td>\n",
       "      <td>0.428038</td>\n",
       "      <td>0.196191</td>\n",
       "      <td>0.307977</td>\n",
       "      <td>0.767870</td>\n",
       "      <td>0.586168</td>\n",
       "      <td>0.434599</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8295</th>\n",
       "      <td>4614</td>\n",
       "      <td>3660593</td>\n",
       "      <td>Q9H3S4_S243T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>243</td>\n",
       "      <td>Q9H3S4</td>\n",
       "      <td>ST</td>\n",
       "      <td>S243T</td>\n",
       "      <td>Polar</td>\n",
       "      <td>...</td>\n",
       "      <td>0.5</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Q9H3S4</td>\n",
       "      <td>1.2412</td>\n",
       "      <td>0.204614</td>\n",
       "      <td>0.108648</td>\n",
       "      <td>0.135211</td>\n",
       "      <td>0.552730</td>\n",
       "      <td>0.356165</td>\n",
       "      <td>0.434599</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8296 rows × 57 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Unnamed: 0        X1           id2 AA1 AA2  position Uniprot WT_Mut  \\\n",
       "0           8323  19082162  P04035_P439A   P   A       439  P04035     PA   \n",
       "1           8329  19082168  P04035_P439H   P   H       439  P04035     PH   \n",
       "2           8332  19082171  P04035_P439L   P   L       439  P04035     PL   \n",
       "3           8336  19082175  P04035_P439R   P   R       439  P04035     PR   \n",
       "4           8337  19082176  P04035_P439S   P   S       439  P04035     PS   \n",
       "...          ...       ...           ...  ..  ..       ...     ...    ...   \n",
       "8291        4610   3660589  Q9H3S4_S243N   S   N       243  Q9H3S4     SN   \n",
       "8292        4613   3660592  Q9H3S4_S243R   S   R       243  Q9H3S4     SR   \n",
       "8293        4613   3660592  Q9H3S4_S243R   S   R       243  Q9H3S4     SR   \n",
       "8294        4613   3660592  Q9H3S4_S243R   S   R       243  Q9H3S4     SR   \n",
       "8295        4614   3660593  Q9H3S4_S243T   S   T       243  Q9H3S4     ST   \n",
       "\n",
       "     Variant AA1_polarity  ... PhyloP_0  PhyloP_1  protein  Q(SASA)  \\\n",
       "0      P439A      Special  ...      4.0       4.0   P04035   1.0323   \n",
       "1      P439H      Special  ...      4.0      -0.5   P04035   1.0323   \n",
       "2      P439L      Special  ...      4.0      -0.5   P04035   1.0323   \n",
       "3      P439R      Special  ...      4.0      -0.5   P04035   1.0323   \n",
       "4      P439S      Special  ...      4.0       4.0   P04035   1.0323   \n",
       "...      ...          ...  ...      ...       ...      ...      ...   \n",
       "8291   S243N        Polar  ...      0.5       4.0   Q9H3S4   1.2412   \n",
       "8292   S243R        Polar  ...      4.0       5.0   Q9H3S4   1.2412   \n",
       "8293   S243R        Polar  ...      2.0       0.5   Q9H3S4   1.2412   \n",
       "8294   S243R        Polar  ...      4.0       5.0   Q9H3S4   1.2412   \n",
       "8295   S243T        Polar  ...      0.5       4.0   Q9H3S4   1.2412   \n",
       "\n",
       "      AA_Cons_MutSig_QSASA_p  AA_Cons_MutSig_p  AA_MutSig_QSASA_p  \\\n",
       "0                   0.618426          0.195662           0.606245   \n",
       "1                   0.098062          0.110471           0.592379   \n",
       "2                   0.226643          0.150819           0.728743   \n",
       "3                   0.145746          0.185445           0.771501   \n",
       "4                   0.387892          0.295273           0.441203   \n",
       "...                      ...               ...                ...   \n",
       "8291                0.052334          0.085340           0.080482   \n",
       "8292                0.349461          0.246390           0.278748   \n",
       "8293                0.248722          0.188052           0.158618   \n",
       "8294                0.428038          0.196191           0.307977   \n",
       "8295                0.204614          0.108648           0.135211   \n",
       "\n",
       "      AA_MutSig_p      AA_p  MutSig_p  \n",
       "0        0.541139  0.606757  0.408901  \n",
       "1        0.700575  0.650552  0.678710  \n",
       "2        0.789619  0.671871  0.712546  \n",
       "3        0.782489  0.686362  0.685450  \n",
       "4        0.385010  0.555850  0.438892  \n",
       "...           ...       ...       ...  \n",
       "8291     0.420180  0.452736  0.297807  \n",
       "8292     0.827931  0.586168  0.474273  \n",
       "8293     0.114509  0.586168  0.427768  \n",
       "8294     0.767870  0.586168  0.434599  \n",
       "8295     0.552730  0.356165  0.434599  \n",
       "\n",
       "[8296 rows x 57 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "envision = pd.read_csv('../Envision/Envision_ready.tsv', sep = \"\\t\")\n",
    "envision['AA_Cons_MutSig_QSASA_p'] = model.predict_proba( X_test )[:, 1]\n",
    "\n",
    "with open('setup/AA_Cons_MutSig/GBM.pickle', 'rb' ) as r:\n",
    "    model = pickle.load( r )\n",
    "envision['AA_Cons_MutSig_p'] = model.predict_proba( X_test[:, 0:57] )[:, 1]\n",
    "\n",
    "with open('setup/AA_MutSig_QSASA/GBM.pickle', 'rb' ) as r:\n",
    "    model = pickle.load( r )\n",
    "envision['AA_MutSig_QSASA_p'] = model.predict_proba( X_test[:, [i for i in range(54)] + [57]] )[:, 1]\n",
    "\n",
    "with open('setup/AA_MutSig/GBM.pickle', 'rb' ) as r:\n",
    "    model = pickle.load( r )\n",
    "envision['AA_MutSig_p'] = model.predict_proba( X_test[:, 0:54] )[:, 1]\n",
    "\n",
    "with open('setup/AA/GBM.pickle', 'rb' ) as r:\n",
    "    model = pickle.load( r )\n",
    "envision['AA_p'] = model.predict_proba( X_test[:, 0:40] )[:, 1]\n",
    "\n",
    "with open('setup/MutSig/GBM.pickle', 'rb' ) as r:\n",
    "    model = pickle.load( r )\n",
    "envision['MutSig_p'] = model.predict_proba( X_test[:, 41:54] )[:, 1]\n",
    "\n",
    "envision"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "envision.to_csv('../Envision/Envision_DMSexppredicted.tsv', sep = \"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Feature importance\n",
    "\n",
    "Use the [SHAP (SHapley Additive exPlanations)](https://github.com/slundberg/shap) library."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# function to perform SHAP analysis and output the explanatory matrix\n",
    "\n",
    "def doSHAP(model_file, data_file, out_file, column_selection, feature_names):\n",
    "    \"\"\"\n",
    "    automatic routine to put through data through model and output SHAP explanatory matrix\n",
    "    arg:\n",
    "      model_file: filepath to sklearn model pickle\n",
    "      data_file: filepath to dataset (X, Y) pickle\n",
    "      out_file: filepath to output CSV where the SHAP matrix will be written out\n",
    "      column_selection: list of column indices to be included in the feature matrix X for interpretation\n",
    "      feature_names: list of feature names. Use as column names of the output matrix\n",
    "    \"\"\"\n",
    "    with open(model_file, 'rb' ) as r:\n",
    "        model = pickle.load( r )\n",
    "    with open(data_file, 'rb') as r:\n",
    "        X, Y, scoreset = pickle.load(r)\n",
    "    X = X[:, column_selection]    \n",
    "    explainer = shap.Explainer(model)\n",
    "    shap_values = explainer(X)\n",
    "    out = pd.DataFrame(shap_values.values, columns = [feature_names[i] for i in column_selection])\n",
    "    out.to_csv(out_file, index=False)\n",
    "    return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# feature names\n",
    "feature_cols = []\n",
    "AAs = list('ACDEFGHIKLMNPQRSTVWY')\n",
    "feature_cols +=  ['WT_AA_' + aa for aa in AAs]\n",
    "feature_cols +=  ['MUT_AA_' + aa for aa in AAs]\n",
    "bs = list('AGCT')\n",
    "feature_cols +=  ['MutSig_-1_' + b for b in bs]\n",
    "feature_cols +=  ['MutSig_+1_' + b for b in bs]\n",
    "subs = ['C>A', 'C>G', 'C>T', 'T>A', 'T>C', 'T>G']\n",
    "feature_cols +=  ['MutSig_' + m for m in subs] \n",
    "feature_cols += ['PhyloP_0', 'PhyloP_-1', 'PhyloP_1', 'Q(SASA)']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# explain the model's predictions using SHAP\n",
    "doSHAP(model_file = 'setup/AA_Cons_MutSig_QSASA/GBM.pickle', data_file = 'data/test_set.pickle', \\\n",
    "       out_file = \"GBM-AA_Cons_MutSig_QSASA_test-set_shap-values.csv\", feature_names = feature_cols, \\\n",
    "       column_selection = [i for i in range(58)])\n",
    "doSHAP(model_file = 'setup/AA_MutSig_QSASA/GBM.pickle', data_file = 'data/test_set.pickle', \\\n",
    "       out_file = \"GBM-AA_MutSig_QSASA_test-set_shap-values.csv\", feature_names = feature_cols, \\\n",
    "       column_selection = [i for i in range(54)] + [57])\n",
    "doSHAP(model_file = 'setup/AA_Cons_MutSig/GBM.pickle', data_file = 'data/test_set.pickle', \\\n",
    "       out_file = \"GBM-AA_Cons_MutSig_test-set_shap-values.csv\", feature_names = feature_cols, \\\n",
    "       column_selection = [i for i in range(57)])\n",
    "doSHAP(model_file = 'setup/AA_MutSig/GBM.pickle', data_file = 'data/test_set.pickle', \\\n",
    "       out_file = \"GBM-AA_MutSig_test-set_shap-values.csv\", feature_names = feature_cols, \\\n",
    "       column_selection = [i for i in range(54)])\n",
    "doSHAP(model_file = 'setup/AA/GBM.pickle', data_file = 'data/test_set.pickle', \\\n",
    "       out_file = \"GBM-AA_test-set_shap-values.csv\", feature_names = feature_cols, \\\n",
    "       column_selection = [i for i in range(40)])\n",
    "doSHAP(model_file = 'setup/MutSig/GBM.pickle', data_file = 'data/test_set.pickle', \\\n",
    "       out_file = \"GBM-MutSig_test-set_shap-values.csv\", feature_names = feature_cols, \\\n",
    "       column_selection = [i for i in range(40, 54)])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
